#used to generate the input for michigan impuation server
#some input files were generated by extrac_cambridgesamples.R (task2)
cd /fh/fast/dai_j/BEACON/BEACON_GRANT/code
ml Python #for snpflip
plink=/fh/fast/stanford_j/Xiaoyu_Oct2020/Tools/plink-1.07-x86_64/plink1.9/plink
vcfsort=/fh/fast/dai_j/CancerGenomics/Tools/vcftools_0.1.12/bin/vcf-sort
snpflip=/fh/fast/dai_j/CancerGenomics/Tools/python3/snpflip/snpflip
reference=/fh/fast/stanford_j/Xiaoyu_Oct2020/Tools/reference/human_g1k_v37.fasta
outfolder=/fh/fast/dai_j/BEACON/BEACON_GRANT/result
#QC on call-rates of snp and sample
$plink --bfile /fh/fast/dai_j/BEACON/BeagessCambridgeAmos/bca_filtered_19Feb2015 --geno 0.1 --mind 0.1 --update-alleles ../result/updateallele.txt --remove ../result/bca_michigan_imp_statistics_lowcallratesamples.txt -make-bed --out $outfolder/bca_filtered_30Nov2018

#exclued AB snps (those didn't update allele coding)
$plink --bfile $outfolder/bca_filtered_30Nov2018 --exclude ../result/excludeABsnps.txt  -make-bed --out $outfolder/bca_filtered_30Nov2018

snps=($(cat '../result/subsetsnp_5M.txt')) #select snps around

chrs=(2 3 5 6 7 8 9 12 15 16 19) #chrs include those 23 snps
n=0
for i in ${chrs[@]}
do
  $plink --bfile $outfolder/bca_filtered_30Nov2018 --chr $i --recode --update-alleles ../result/updateallele.txt --snps ${snps[$n]} -make-bed --out $outfolder/bca_filtered_30Nov2018_chr$i
  ((n=n+1))
  echo $n
  #flip snp strand
  $snpflip -b $outfolder/bca_filtered_30Nov2018_chr$i.bim -f $reference -o $outfolder/snpfilp_output
  $plink --bfile $outfolder/bca_filtered_30Nov2018_chr$i --flip $outfolder/snpfilp_output.reverse -make-bed --out $outfolder/bca_filtered_30Nov2018_chr$i

  $plink --bfile $outfolder/bca_filtered_30Nov2018_chr$i --recode vcf --out $outfolder/bca_filtered_30Nov2018_chr$i
  
  $vcfsort $outfolder/bca_filtered_30Nov2018_chr$i.vcf |bgzip -c >$outfolder/bca_filtered_30Nov2018_chr$i.vcf.gz
done

#for all genotypes

for i in {1..22}
do
  $plink --bfile $outfolder/bca_filtered_30Nov2018 --chr $i  --mind 0.05 --geno 0.05 --hwe 0.000001 --maf 0.05 --remove ../result/bca_michigan_imp_removedsamples.txt --recode --update-alleles ../result/updateallele.txt -make-bed --out $outfolder/bca_filtered_19Dec2018_chr$i #remove samples in the second time once michigan found mismatchs for samples in different chrs
  $snpflip -b $outfolder/bca_filtered_19Dec2018_chr$i.bim -f $reference -o $outfolder/snpfilp_output
  $plink --bfile $outfolder/bca_filtered_19Dec2018_chr$i --flip $outfolder/snpfilp_output.reverse -make-bed --out $outfolder/bca_filtered_19Dec2018_chr$i
  $plink --bfile $outfolder/bca_filtered_19Dec2018_chr$i --recode vcf --out $outfolder/bca_filtered_19Dec2018_chr$i
  $vcfsort $outfolder/bca_filtered_19Dec2018_chr$i.vcf |bgzip -c >$outfolder/bca_filtered_19Dec2018_chr$i.vcf.gz
done
i=23
$plink --bfile $outfolder/bca_filtered_30Nov2018 --chr $i --list-duplicate-vars
$plink --bfile $outfolder/bca_filtered_30Nov2018 --chr $i  --mind 0.05 --geno 0.05 --hwe 0.000001 --maf 0.05 --remove ../result/imputation_vcf/bca_michigan_imp_removedsamples.txt --recode --update-alleles ../result/updateallele.txt --exclude ./plink.dupvar -make-bed --out $outfolder/bca_filtered_19Dec2018_chr$i
#sed 's/^23/X/' $outfolder/bca_filtered_19Dec2018_chr$i.bim >$outfolder/bca_filtered_19Dec2018_chr$i.bim1
#mv $outfolder/bca_filtered_19Dec2018_chr$i.bim1 $outfolder/bca_filtered_19Dec2018_chr$i.bim
#$snpflip -b $outfolder/bca_filtered_19Dec2018_chr$i.bim -f $reference -o $outfolder/snpfilp_output
#$plink --bfile $outfolder/bca_filtered_19Dec2018_chr$i --list-duplicate-vars
#$plink --bfile $outfolder/bca_filtered_19Dec2018_chr$i --flip $outfolder/snpfilp_output.reverse -make-bed --out $outfolder/bca_filtered_19Dec2018_chr$i
$plink --bfile $outfolder/bca_filtered_19Dec2018_chr$i --recode vcf --set-hh-missing --out $outfolder/bca_filtered_19Dec2018_chr$i #--set-hh-missing only for chr23
sed 's/^23/X/' $outfolder/bca_filtered_19Dec2018_chr$i.vcf > $outfolder/bca_filtered_19Dec2018_chr$i.vcf1
mv $outfolder/bca_filtered_19Dec2018_chr$i.vcf1 $outfolder/bca_filtered_19Dec2018_chr$i.vcf
$vcfsort $outfolder/bca_filtered_19Dec2018_chr$i.vcf |bgzip -c >$outfolder/bca_filtered_19Dec2018_chr$i.vcf.gz

#work on new control data
#new snpflip
#cd /fh/fast/dai_j/CancerGenomics/Tools/snpflip
#python -m venv env
#source ./env/bin/activate
#pip install snpflip
#deactivate

snpflip=/fh/fast/dai_j/CancerGenomics/Tools/snpflip/env/bin/snpflip
bgzip=/fh/fast/stanford_j/Xiaoyu_Oct2020/Tools/ensembl-vep/reference/htslib/bgzip
outfolder=/fh/fast/dai_j/BEACON/BEACON_GRANT/result/imputation_vcf

#BEACON_dbgapcontrol_newID was generated in read_controls.R
$plink --bfile /fh/fast/dai_j/BEACON/BEACON_GRANT/data/AdditionalControlPuya/BEACON_dbgapcontrol_newID --keep ../result/BEACON_dbgapcontrol_plinksamples.txt  -make-bed --out $outfolder/BEACON_dbgapcontrol

#check missingness
prefix=/fh/fast/dai_j/BEACON/BEACON_GRANT/data/AdditionalControlPuya/BEACON_dbgapcontrol_newID
prefix1=BEACON_dbgapcontrol_newID
checkmissing(){
  local chr="$1"
  echo $chr
  $plink --bfile $prefix --chr $chr --missing --out $outfolder/${prefix1}_chr${chr}_missing --memory 2560
}
for chr in {1..22}
do
 checkmissing $chr &
done

#need to apply QC on each chr, or chr17 not imputed for beacon data (multiple chunk(s) excluded: at least one sample has a call rate < 50.0%)
for i in {1..22}
do
  $plink --bfile /fh/fast/dai_j/BEACON/BEACON_GRANT/data/AdditionalControlPuya/BEACON_dbgapcontrol_newID  --chr $i --mind 0.05 --geno 0.05 --hwe 0.000001 -make-bed --out $outfolder/BEACON_dbgapcontrol_chr$i 
  $snpflip -b $outfolder/BEACON_dbgapcontrol_chr$i.bim -f $reference -o $outfolder/snpfilp_output
  $plink --bfile $outfolder/BEACON_dbgapcontrol_chr$i --flip $outfolder/snpfilp_output.reverse -make-bed --out $outfolder/BEACON_dbgapcontrol_chr$i
  $plink --bfile $outfolder/BEACON_dbgapcontrol_chr$i --recode vcf --out $outfolder/BEACON_dbgapcontrol_chr$i
  $vcfsort $outfolder/BEACON_dbgapcontrol_chr$i.vcf |$bgzip -c >$outfolder/BEACON_dbgapcontrol_chr$i.vcf.gz
done
#number of samples are not same accross chr
for i in {1..22}
do
  $plink --bfile /fh/fast/dai_j/BEACON/BEACON_GRANT/data/AdditionalControlPuya/BEACON_dbgapcontrol_newID  --chr $i --mind 0.05 --geno 0.05 --hwe 0.000001 --keep ../result/BEACON_dbgapcontrol_QCplinksamples.txt -make-bed --out $outfolder/BEACON_dbgapcontrol_chr$i 
  $snpflip -b $outfolder/BEACON_dbgapcontrol_chr$i.bim -f $reference -o $outfolder/snpfilp_output
  $plink --bfile $outfolder/BEACON_dbgapcontrol_chr$i --flip $outfolder/snpfilp_output.reverse -make-bed --out $outfolder/BEACON_dbgapcontrol_chr$i
  $plink --bfile $outfolder/BEACON_dbgapcontrol_chr$i --recode vcf --out $outfolder/BEACON_dbgapcontrol_chr$i
  $vcfsort $outfolder/BEACON_dbgapcontrol_chr$i.vcf |$bgzip -c >$outfolder/BEACON_dbgapcontrol_chr$i.vcf.gz
done


$plink --bfile /fh/fast/dai_j/BEACON/BEACON_GRANT/data/AdditionalControlPuya/Cambridge_WTCCC_newID --keep ../result/Cambridge_WTCCC_plinksamples.txt  -make-bed --out $outfolder/Cambridge_WTCCC
prefix=$outfolder/Cambridge_WTCCC
prefix1=Cambridge_WTCCC
for chr in {1..22}
do
 checkmissing $chr &
done

for i in {1..22}
do
  $plink --bfile /fh/fast/dai_j/BEACON/BEACON_GRANT/data/AdditionalControlPuya/Cambridge_WTCCC_newID  --chr $i --mind 0.05 --geno 0.05 --hwe 0.000001 --keep ../result/Cambridge_WTCCC_plinksamples.txt  -make-bed --out $outfolder/Cambridge_WTCCCcontrol_chr$i 
  $snpflip -b $outfolder/Cambridge_WTCCCcontrol_chr$i.bim -f $reference -o $outfolder/snpfilp_output
  $plink --bfile $outfolder/Cambridge_WTCCCcontrol_chr$i --flip $outfolder/snpfilp_output.reverse -make-bed --out $outfolder/Cambridge_WTCCCcontrol_chr$i
  $plink --bfile $outfolder/Cambridge_WTCCCcontrol_chr$i --recode vcf --out $outfolder/Cambridge_WTCCCcontrol_chr$i
  $vcfsort $outfolder/Cambridge_WTCCCcontrol_chr$i.vcf |$bgzip -c >$outfolder/Cambridge_WTCCCcontrol_chr$i.vcf.gz
done
for i in {1..22}
do
  $plink --bfile /fh/fast/dai_j/BEACON/BEACON_GRANT/data/AdditionalControlPuya/Cambridge_WTCCC_newID  --chr $i --mind 0.05 --geno 0.05 --hwe 0.000001 --keep ../result/Cambridge_WTCCC_QCplinksamples.txt  -make-bed --out $outfolder/Cambridge_WTCCCcontrol_chr$i 
  $snpflip -b $outfolder/Cambridge_WTCCCcontrol_chr$i.bim -f $reference -o $outfolder/snpfilp_output
  $plink --bfile $outfolder/Cambridge_WTCCCcontrol_chr$i --flip $outfolder/snpfilp_output.reverse -make-bed --out $outfolder/Cambridge_WTCCCcontrol_chr$i
  $plink --bfile $outfolder/Cambridge_WTCCCcontrol_chr$i --recode vcf --out $outfolder/Cambridge_WTCCCcontrol_chr$i
  $vcfsort $outfolder/Cambridge_WTCCCcontrol_chr$i.vcf |$bgzip -c >$outfolder/Cambridge_WTCCCcontrol_chr$i.vcf.gz
done

#after download results, they are unziped using data/imputation/unzip_imputationres.sh

#add missing gwas snps "rs9918259"  "rs75783973" in chr5, 1000genome phase1
#submit BEACON_dbgapcontrol_chr5.vcf.gz, Cambridge_WTCCCcontrol_chr5.vcf.gz to michigan server again (Chr5_Beacon_phase1,Chr5_Cambridge_phase1)