#the code generate covariates/phenotype using samples filtered by GCTA_CO_BEEA_rm05.grm.id (generated by extract_genotype_GCTA1.sh), outputs: *.pheno1,*.covar1
#remove relatedness
rm(list=ls())
if (!exists("sampletable"))
{
  #library(xlsx)
  #sampletable=read.xlsx("/fh/fast/dai_j/BEACON/BeagessCambridgeAmos/bc_AT_JD1.xlsx",sheetIndex = 1,stringsAsFactors=F)
  sampletable=read.table("/fh/fast/dai_j/BEACON/BEACON_GRANT/data//bc_AT_JD1.csv",header=T,sep=",",stringsAsFactors=F)
  for (i in 1:ncol(sampletable))
  {
    idx=which(sampletable[,i]==-9)
    sampletable[idx,i]=NA
  }
}

filteredtable=read.table("../result/GCTA_CO_BEEA_rm05.grm.id",stringsAsFactors = F)
fam=read.table("../result/bca_filtered_04Apr2019.fam",stringsAsFactors = F)
sum(sampletable$phenoBE_bc==2 & sampletable$phenoEA_bc==2,na.rm=T) #0
sum(fam$V2 %in% sampletable$localid) #8208
sum(fam$V2 %in% filteredtable$V2) #7829

fam=fam[fam$V2 %in% filteredtable$V2,]
all(sampletable$localid %in% fam$V2) #F
all(fam$V2 %in% sampletable$localid) #T
pheno=data.frame(famid=fam$V1,subjid=fam$V2,pheno=-9,stringsAsFactors = F)
idx1=which(pheno$subjid %in% sampletable$localid[sampletable$phenoBE_bc==1 | sampletable$phenoEA_bc==1]) #control
idx2=which(pheno$subjid %in% sampletable$localid[sampletable$phenoBE_bc==2]) #BE case
idx3=which(pheno$subjid %in% sampletable$localid[sampletable$phenoEA_bc==2]) #EA case
length(intersect(idx1,idx2))
length(intersect(idx3,idx2))
length(intersect(idx1,idx3))
pheno$pheno[idx1]=0
pheno$pheno[c(idx2,idx3)]=1
table(pheno$pheno)
# 0    1 
# 2106 5723 
write.table(pheno,file="../result/bca_filtered_04Apr2019.GCTA.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[c(idx1,idx2),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BE.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[c(idx1,idx3),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.EA.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[c(idx1,idx2,idx3),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
pheno$EA=-9
pheno$EA[c(idx1,idx2)]=0
pheno$EA[idx3]=1
write.table(pheno[c(idx2,idx3),c(1,2,4)],file="../result/bca_filtered_04Apr2019.GCTA.BE.EA.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[c(idx1,idx2),1:2],file="../result/GCTA_CO_BE_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[c(idx1,idx3),1:2],file="../result/GCTA_CO_EA_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[c(idx2,idx3),1:2],file="../result/GCTA_BE_EA_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[c(idx1,idx2,idx3),1:2],file="../result/GCTA_CO_BEEA_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
# #used for plink to filter samples
# tmp=pheno[c(idx1,idx2,idx3),]
# tmp=data.frame(fam=rep(0,length(idx1)+length(idx2)+length(idx3)),ind=paste0(tmp$famid,"_",tmp$subjid),stringsAsFactors = F)
# write.table(tmp,file="../result/GCTA_CO_BEEA_individuals1_plink.txt",col.names = F,row.names = F,quote=F)
#subseting recurrent_HB_RF
idx=match(pheno$subjid,sampletable$localid)
pheno$recurrent_HB_RF=NA
pheno$recurrent_HB_RF=sampletable$recurrent_HB_RF[idx]
idx4=which(pheno$recurrent_HB_RF==1)
write.table(pheno[intersect(c(idx1,idx2),idx4),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BE.recurrent1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx4),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.EA.recurrent1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx4),c(1,2,4)],file="../result/bca_filtered_04Apr2019.GCTA.BE.EA.recurrent1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx4),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.recurrent1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2),idx4),1:2],file="../result/GCTA_CO_BE_recurrent1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx4),1:2],file="../result/GCTA_CO_EA_recurrent1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx4),1:2],file="../result/GCTA_BE_EA_recurrent1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx4),1:2],file="../result/GCTA_CO_BEEA_recurrent1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
idx4=which(pheno$recurrent_HB_RF==0)
write.table(pheno[intersect(c(idx1,idx2),idx4),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BE.recurrent0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx4),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.EA.recurrent0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx4),c(1,2,4)],file="../result/bca_filtered_04Apr2019.GCTA.BE.EA.recurrent0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx4),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.recurrent0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2),idx4),1:2],file="../result/GCTA_CO_BE_recurrent0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx4),1:2],file="../result/GCTA_CO_EA_recurrent0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx4),1:2],file="../result/GCTA_BE_EA_recurrent0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx4),1:2],file="../result/GCTA_CO_BEEA_recurrent0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
#subseting smoking
idx=match(pheno$subjid,sampletable$localid)
pheno$cig_smk_ever=NA
pheno$cig_smk_ever=sampletable$cig_smk_ever[idx]
idx5=which(pheno$cig_smk_ever==1)
write.table(pheno[intersect(c(idx1,idx2),idx5),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BE.smoke1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx5),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.EA.smoke1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx5),c(1,2,4)],file="../result/bca_filtered_04Apr2019.GCTA.BE.EA.smoke1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx5),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.smoke1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2),idx5),1:2],file="../result/GCTA_CO_BE_smoke1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx5),1:2],file="../result/GCTA_CO_EA_smoke1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx5),1:2],file="../result/GCTA_BE_EA_smoke1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx5),1:2],file="../result/GCTA_CO_BEEA_smoke1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
idx5=which(pheno$cig_smk_ever==0)
write.table(pheno[intersect(c(idx1,idx2),idx5),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BE.smoke0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx5),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.EA.smoke0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx5),c(1,2,4)],file="../result/bca_filtered_04Apr2019.GCTA.BE.EA.smoke0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx5),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.smoke0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2),idx5),1:2],file="../result/GCTA_CO_BE_smoke0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx5),1:2],file="../result/GCTA_CO_EA_smoke0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx5),1:2],file="../result/GCTA_BE_EA_smoke0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx5),1:2],file="../result/GCTA_CO_BEEA_smoke0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
#subseting bmi
idx=match(pheno$subjid,sampletable$localid)
pheno$bmi_recent_healthy=NA
pheno$bmi_recent_healthy=sampletable$bmi_recent_healthy[idx]
idx6=which(pheno$bmi_recent_healthy<=27.37)
write.table(pheno[intersect(c(idx1,idx2),idx6),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BE.bmi0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx6),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.EA.bmi0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx6),c(1,2,4)],file="../result/bca_filtered_04Apr2019.GCTA.BE.EA.bmi0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx6),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.bmi0.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2),idx6),1:2],file="../result/GCTA_CO_BE_bmi0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx6),1:2],file="../result/GCTA_CO_EA_bmi0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx6),1:2],file="../result/GCTA_BE_EA_bmi0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx6),1:2],file="../result/GCTA_CO_BEEA_bmi0_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
idx6=which(pheno$bmi_recent_healthy>27.37)
write.table(pheno[intersect(c(idx1,idx2),idx6),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BE.bmi1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx6),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.EA.bmi1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx6),c(1,2,4)],file="../result/bca_filtered_04Apr2019.GCTA.BE.EA.bmi1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx6),1:3],file="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.bmi1.pheno1",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2),idx6),1:2],file="../result/GCTA_CO_BE_bmi1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx3),idx6),1:2],file="../result/GCTA_CO_EA_bmi1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx2,idx3),idx6),1:2],file="../result/GCTA_BE_EA_bmi1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)
write.table(pheno[intersect(c(idx1,idx2,idx3),idx6),1:2],file="../result/GCTA_CO_BEEA_bmi1_individuals1.txt",sep=" ",col.names = F,row.names = F,quote=F)


#generate the cov input,sex,site
generate_cov=function(idx=c(idx1,idx2),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.covar1")
{
  res=pheno[idx,1:2]
  idxsampletable=match(pheno$subjid[idx],sampletable$localid)
  if (sum(is.na(idxsampletable))>0) warning("some ids are not mapped")
  res$sex=sampletable$sex[idxsampletable]
  res$sex[res$sex==1]="Male"
  res$sex[res$sex==2]="Female"
  # res$site=sampletable$site[idxsampletable]
  # if (sum(is.na(res$site))>0) warning("some sites are missing")
  # res$site=paste0("Site",res$site)
  write.table(res,file=outfile,sep=" ",row.names = F,col.names = F,quote=F)
}
generate_cov()
generate_cov(idx=c(idx1,idx3),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.covar1")
generate_cov(idx=c(idx2,idx3),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.covar1")
generate_cov(idx=c(idx1,idx2,idx3),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.covar1")
#subseting recurrent_HB_RF
idx4=which(pheno$recurrent_HB_RF==1)
generate_cov(idx=intersect(c(idx1,idx2),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.recurrent1.covar1")
generate_cov(idx=intersect(c(idx1,idx3),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.recurrent1.covar1")
generate_cov(idx=intersect(c(idx2,idx3),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.recurrent1.covar1")
generate_cov(idx=intersect(c(idx1,idx2,idx3),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.recurrent1.covar1")
idx4=which(pheno$recurrent_HB_RF==0)
generate_cov(idx=intersect(c(idx1,idx2),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.recurrent0.covar1")
generate_cov(idx=intersect(c(idx1,idx3),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.recurrent0.covar1")
generate_cov(idx=intersect(c(idx2,idx3),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.recurrent0.covar1")
generate_cov(idx=intersect(c(idx1,idx2,idx3),idx4),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.recurrent0.covar1")
idx5=which(pheno$cig_smk_ever==1)
generate_cov(idx=intersect(c(idx1,idx2),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.smoke1.covar1")
generate_cov(idx=intersect(c(idx1,idx3),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.smoke1.covar1")
generate_cov(idx=intersect(c(idx2,idx3),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.smoke1.covar1")
generate_cov(idx=intersect(c(idx1,idx2,idx3),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.smoke1.covar1")
idx5=which(pheno$cig_smk_ever==0)
generate_cov(idx=intersect(c(idx1,idx2),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.smoke0.covar1")
generate_cov(idx=intersect(c(idx1,idx3),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.smoke0.covar1")
generate_cov(idx=intersect(c(idx2,idx3),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.smoke0.covar1")
generate_cov(idx=intersect(c(idx1,idx2,idx3),idx5),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.smoke0.covar1")
idx6=which(pheno$bmi_recent_healthy<=27.37)
generate_cov(idx=intersect(c(idx1,idx2),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.bmi0.covar1")
generate_cov(idx=intersect(c(idx1,idx3),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.bmi0.covar1")
generate_cov(idx=intersect(c(idx2,idx3),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.bmi0.covar1")
generate_cov(idx=intersect(c(idx1,idx2,idx3),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.bmi0.covar1")
idx6=which(pheno$bmi_recent_healthy>27.37)
generate_cov(idx=intersect(c(idx1,idx2),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.bmi1.covar1")
generate_cov(idx=intersect(c(idx1,idx3),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.bmi1.covar1")
generate_cov(idx=intersect(c(idx2,idx3),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.bmi1.covar1")
generate_cov(idx=intersect(c(idx1,idx2,idx3),idx6),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.bmi1.covar1")



#use recurrent as covariate
generate_cov1=function(idx=c(idx1,idx2),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.recurrent.covar1",opt="recurrent")
{
  res=pheno[idx,1:2]
  idxsampletable=match(pheno$subjid[idx],sampletable$localid)
  if (sum(is.na(idxsampletable))>0) warning("some ids are not mapped")
  res$sex=sampletable$sex[idxsampletable]
  res$sex[res$sex==1]="Male"
  res$sex[res$sex==2]="Female"
  if (opt=="recurrent")
  {
    res$recurrent_HB_RF=sampletable$recurrent_HB_RF[idxsampletable]
    res$recurrent_HB_RF[res$recurrent_HB_RF==0]="recurrent_HB_RF0"
    res$recurrent_HB_RF[res$recurrent_HB_RF==1]="recurrent_HB_RF1"
  }
  if (opt=="smoke")
  {
    res$cig_smk_ever=sampletable$cig_smk_ever[idxsampletable]
    res$cig_smk_ever[res$cig_smk_ever==0]="cig_smk_ever0"
    res$cig_smk_ever[res$cig_smk_ever==1]="recurrent_HB_RF1"
  }
  
  # res$site=sampletable$site[idxsampletable]
  # if (sum(is.na(res$site))>0) warning("some sites are missing")
  # res$site=paste0("Site",res$site)
  write.table(res,file=outfile,sep=" ",row.names = F,col.names = F,quote=F)
}
generate_cov1()
generate_cov1(idx=c(idx1,idx3),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.recurrent.covar1")
generate_cov1(idx=c(idx2,idx3),outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.recurrent.covar1")
generate_cov1(idx=c(idx1,idx2,idx3),outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.recurrent.covar1")

generate_qcov=function(idx=c(idx1,idx2),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.qcovar1")
{
  res=pheno[idx,1:2]
  pcadat=read.table(pcafile,stringsAsFactors = F)
  idxa=match(pheno$subjid[idx],pcadat$V2)
  pcadat=pcadat[idxa,]
  res=cbind(res,pcadat[,3:ncol(pcadat)])
  idxsampletable=match(pheno$subjid[idx],sampletable$localid)
  res$age=sampletable$age[idxsampletable]
  write.table(res,file=outfile,sep=" ",row.names = F,col.names = F,quote=F)
}

generate_qcov(idx=c(idx1,idx2),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.qcovar1")
generate_qcov(idx=c(idx1,idx3),pcafile="../result/GCTA_CO_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.qcovar1")
generate_qcov(idx=c(idx2,idx3),pcafile="../result/GCTA_BE_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.qcovar1")
generate_qcov(idx=c(idx1,idx2,idx3),pcafile="../result/GCTA_CO_BEEA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.qcovar1")

#subseting recurrent_HB_RF
idx4=which(pheno$recurrent_HB_RF==1)
generate_qcov(idx=intersect(c(idx1,idx2),idx4),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.recurrent1.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx3),idx4),pcafile="../result/GCTA_CO_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.recurrent1.qcovar1")
generate_qcov(idx=intersect(c(idx2,idx3),idx4),pcafile="../result/GCTA_BE_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.recurrent1.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx2,idx3),idx4),pcafile="../result/GCTA_CO_BEEA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.recurrent1.qcovar1")
idx4=which(pheno$recurrent_HB_RF==0)
generate_qcov(idx=intersect(c(idx1,idx2),idx4),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.recurrent0.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx3),idx4),pcafile="../result/GCTA_CO_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.recurrent0.qcovar1")
generate_qcov(idx=intersect(c(idx2,idx3),idx4),pcafile="../result/GCTA_BE_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.recurrent0.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx2,idx3),idx4),pcafile="../result/GCTA_CO_BEEA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.recurrent0.qcovar1")
idx5=which(pheno$cig_smk_ever==0)
generate_qcov(idx=intersect(c(idx1,idx2),idx5),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.smoke0.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx3),idx5),pcafile="../result/GCTA_CO_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.smoke0.qcovar1")
generate_qcov(idx=intersect(c(idx2,idx3),idx5),pcafile="../result/GCTA_BE_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.smoke0.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx2,idx3),idx5),pcafile="../result/GCTA_CO_BEEA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.smoke0.qcovar1")
idx5=which(pheno$cig_smk_ever==1)
generate_qcov(idx=intersect(c(idx1,idx2),idx5),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.smoke1.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx3),idx5),pcafile="../result/GCTA_CO_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.smoke1.qcovar1")
generate_qcov(idx=intersect(c(idx2,idx3),idx5),pcafile="../result/GCTA_BE_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.smoke1.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx2,idx3),idx5),pcafile="../result/GCTA_CO_BEEA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.smoke1.qcovar1")
idx6=which(pheno$bmi_recent_healthy<=27.37)
generate_qcov(idx=intersect(c(idx1,idx2),idx6),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.bmi0.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx3),idx6),pcafile="../result/GCTA_CO_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.bmi0.qcovar1")
generate_qcov(idx=intersect(c(idx2,idx3),idx6),pcafile="../result/GCTA_BE_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.bmi0.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx2,idx3),idx6),pcafile="../result/GCTA_CO_BEEA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.bmi0.qcovar1")
idx6=which(pheno$bmi_recent_healthy>27.37)
generate_qcov(idx=intersect(c(idx1,idx2),idx6),pcafile="../result/GCTA_CO_BE.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.bmi1.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx3),idx6),pcafile="../result/GCTA_CO_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.EA.bmi1.qcovar1")
generate_qcov(idx=intersect(c(idx2,idx3),idx6),pcafile="../result/GCTA_BE_EA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.BE.EA.bmi1.qcovar1")
generate_qcov(idx=intersect(c(idx1,idx2,idx3),idx6),pcafile="../result/GCTA_CO_BEEA.eigenvec",outfile="../result/bca_filtered_04Apr2019.GCTA.CO.BEEA.bmi1.qcovar1")


# #stratify the SNPs by LD scores of individual SNPs in R
# #https://cnsgenomics.com/software/gcta/#GREMLinWGSorimputeddata
# library(data.table)
# lds_seg = fread("../result/test.score.ld",header=T)
# quartiles=summary(lds_seg$ldscore_SNP)
# 
# lb1 = which(lds_seg$ldscore_SNP <= quartiles[2])
# lb2 = which(lds_seg$ldscore_SNP > quartiles[2] & lds_seg$ldscore_SNP <= quartiles[3])
# lb3 = which(lds_seg$ldscore_SNP > quartiles[3] & lds_seg$ldscore_SNP <= quartiles[5])
# lb4 = which(lds_seg$ldscore_SNP > quartiles[5])
# 
# lb1_snp = data.frame(snp=lds_seg$SNP[lb1])
# lb2_snp = data.frame(snp=lds_seg$SNP[lb2])
# lb3_snp = data.frame(snp=lds_seg$SNP[lb3])
# lb4_snp = data.frame(snp=lds_seg$SNP[lb4])
# 
# fwrite(lb1_snp, "../result/snp_group1.txt", row.names=F, quote=F, col.names=F)
# fwrite(lb2_snp, "../result/snp_group2.txt", row.names=F, quote=F, col.names=F)
# fwrite(lb3_snp, "../result/snp_group3.txt", row.names=F, quote=F, col.names=F)
# fwrite(lb4_snp, "../result/snp_group4.txt", row.names=F, quote=F, col.names=F)


check_covariate=function(y=c(rep(0,length(idx1)),rep(1,length(idx2))),covfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.covar",qcovfile="../result/bca_filtered_04Apr2019.GCTA.CO.BE.qcovar")
{
  covar=read.table(covfile,stringsAsFactors = F)
  qcovar=read.table(qcovfile,stringsAsFactors = F)
  tmp=cbind(covar[,3:ncol(covar)],qcovar[,3:ncol(qcovar)])
  colnames(tmp)=c("Sex","Site",paste0("PC",1:20),"Age")
  fit=glm(y~.,data=tmp)
  tmp1=model.matrix(~.,data=tmp)
  
}

readhsq=function(hsqfile="../result/GCTA_CO_BE_covar.hsq",opt="1grm")
{
  tmp=read.table(hsqfile,sep="\t",fill=T,stringsAsFactors = F,header = T)
  if (opt=="1grm")
  {
    tmp1=data.frame(H=tmp$Variance[c(4,7)],SE=tmp$SE[c(4,7)],pvalue=tmp$Variance[12],stringsAsFactors = F)
  }else #multiple grms
  {
    idx=which(grepl("Sum",tmp[,1]))
    tmp1=data.frame(H=tmp$Variance[idx],SE=tmp$SE[idx],pvalue=tmp$Variance[nrow(tmp)-1],stringsAsFactors = F)
  }
  return(tmp1)
}
readhsq()
# H       SE pvalue
# 1 0.564408 0.062474      0
# 2 0.364057 0.040298      0
readhsq(hsqfile="../result/GCTA_CO_EA_covar.hsq")
# H       SE pvalue
# 1 0.618551 0.072198      0
# 2 0.256594 0.029950      0
readhsq(hsqfile="../result/GCTA_BE_EA_covar.hsq")
# H       SE    pvalue
# 1 0.156066 0.059618 0.0034029
# 2 0.087784 0.033534 0.0034029
readhsq(hsqfile="../result/GCTA_CO_BEEA_covar.hsq")
# H       SE pvalue
# 1 0.462828 0.043705      0
# 2 0.359769 0.033973      0
readhsq(hsqfile="../result/GCTA_CO_BE_covar_recurrent1.hsq")
# H       SE   pvalue
#1 0.325742 0.218424 0.061372
#2 0.286741 0.192272 0.061372
## 1 0.378872 0.195269 0.023097
readhsq(hsqfile="../result/GCTA_CO_BE_covar_recurrent0.hsq")
# H       SE     pvalue
# 1 0.531605 0.137236 3.9806e-05
# 2 0.337186 0.087046 3.9806e-05
readhsq(hsqfile="../result/GCTA_CO_EA_covar_recurrent1.hsq")
# H       SE     pvalue
# 1 0.884951 0.278931 0.00077176
# 2 0.442589 0.139501 0.00077176
readhsq(hsqfile="../result/GCTA_CO_EA_covar_recurrent0.hsq")
# H       SE     pvalue
# 1 0.60361 0.138062 2.3584e-06
# 2 0.26003 0.059476 2.3584e-06
readhsq(hsqfile="../result/GCTA_BE_EA_covar_recurrent1.hsq")
# H       SE  pvalue
# 1 0.044969 0.09361 0.31198
readhsq(hsqfile="../result/GCTA_BE_EA_covar_recurrent0.hsq")
# H       SE    pvalue
# 1 0.22888 0.095639 0.0081457
readhsq(hsqfile="../result/GCTA_CO_BEEA_covar_recurrent1.hsq")
# H       SE    pvalue
# 1 0.495421 0.178913 0.0025483
readhsq(hsqfile="../result/GCTA_CO_BEEA_covar_recurrent0.hsq")
# H       SE     pvalue
# 1 0.334252 0.062528 1.2708e-08
readhsq(hsqfile="../result/GCTA_CO_BE_covar_recurrent.hsq")
# H       SE     pvalue
# 1 0.269713 0.053639 1.1938e-07
readhsq(hsqfile="../result/GCTA_CO_EA_covar_recurrent.hsq")
# H       SE     pvalue
# 1 0.214692 0.038196 1.2717e-09
readhsq(hsqfile="../result/GCTA_BE_EA_covar_recurrent.hsq")
# H       SE    pvalue
# 1 0.107966 0.046573 0.0079086
readhsq(hsqfile="../result/GCTA_CO_BEEA_covar_recurrent.hsq")
# H       SE   pvalue
# 1 0.272575 0.042966 1.01e-11
readhsq(hsqfile="../result/GCTA_CO_BEEA_genotyped.hsq")
# H       SE pvalue
# 1 0.468792 0.044611      0
# 2 0.368021 0.035021      0
readhsq(hsqfile="../result/GCTA_CO_BEEA_recurrent0_genotyped.hsq")
# H       SE     pvalue
# 1 0.519238 0.100095 3.2398e-08
# 2 0.330588 0.063729 3.2398e-08
readhsq(hsqfile="../result/GCTA_CO_BEEA_recurrent1_genotyped.hsq")
# H       SE    pvalue
# 1 0.366585 0.146468 0.0054435
# 2 0.460704 0.184073 0.0054435
readhsq(hsqfile="../result/GCTA_CO_BEEA_smoke0_genotyped.hsq")
# H       SE     pvalue
# 1 0.571618 0.138913 1.4372e-05
# 2 0.390410 0.094876 1.4372e-05
readhsq(hsqfile="../result/GCTA_CO_BEEA_smoke1_genotyped.hsq")
# H       SE     pvalue
# 1 0.547845 0.070802 5.5511e-16
# 2 0.446099 0.057653 5.5511e-16
readhsq(hsqfile="../result/GCTA_CO_BEEA_bmi0_genotyped.hsq")
# H       SE     pvalue
# 1 0.557289 0.151118 0.00010077
# 2 0.347442 0.094215 0.00010077
readhsq(hsqfile="../result/GCTA_CO_BEEA_bmi1_genotyped.hsq")
# H       SE     pvalue
# 1 0.453764 0.099133 1.2707e-06
# 2 0.332709 0.072686 1.2707e-06

#use imputation result
readhsq(hsqfile="../result/imp_CO_BEEA.hsq",opt="mgrm")
# H       SE     pvalue
# 1 0.640763 0.066868 1.9436e-05
# 2 0.503251 0.052518 1.9436e-05
readhsq(hsqfile="../result/imp_CO_BEEA_recurrent1.hsq",opt="mgrm")
# H       SE    pvalue
# 1 0.883743 0.167510 0.0036963
# 2 1.112204 0.210814 0.0036963
readhsq(hsqfile="../result/imp_CO_BEEA_recurrent0.hsq",opt="mgrm")
# H      SE  pvalue
# 1 0.447345 0.12097 0.39225
# 2 0.284854 0.07703 0.39225
readhsq(hsqfile="../result/imp_CO_BE.hsq",opt="mgrm")
# H       SE   pvalue
# 1 0.513875 0.084624 0.012161
# 2 0.332823 0.054809 0.012161
readhsq(hsqfile="../result/imp_CO_BE_recurrent1.hsq",opt="mgrm")
# H       SE  pvalue
# 1 0.425376 0.229453 0.32113
# 2 0.378234 0.204024 0.32113
readhsq(hsqfile="../result/imp_CO_BE_recurrent0.hsq",opt="mgrm")
# H       SE  pvalue
# 1 0.278977 0.146429 0.38981
# 2 0.176935 0.092869 0.38981
readhsq(hsqfile="../result/imp_CO_EA.hsq",opt="mgrm")
# H       SE   pvalue
# 1 0.528353 0.096221 0.019788
# 2 0.219597 0.039992 0.019788
readhsq(hsqfile="../result/imp_CO_EA_recurrent1.hsq",opt="mgrm")
# H       SE  pvalue
# 1 0.999999 0.293622 0.28255
# 2 0.506713 0.148782 0.28255
readhsq(hsqfile="../result/imp_CO_EA_recurrent0.hsq",opt="mgrm")
# 1 0.228864 0.149461 0.48415
# 2 0.098277 0.064180 0.48415

readhsq(hsqfile="../result/imp1grm_CO_BEEA.hsq")
# H       SE pvalue
# 1 0.540965 0.059518      0
# 2 0.424870 0.046745      0
readhsq(hsqfile="../result/imp1grm_CO_BEEA_recurrent1.hsq")
# H       SE     pvalue
# 1 0.708610 0.148199 3.4145e-08
# 2 0.891796 0.186511 3.4145e-08
readhsq(hsqfile="../result/imp1grm_CO_BEEA_recurrent0.hsq")
# H       SE     pvalue
# 1 0.406669 0.112136 8.7901e-05
# 2 0.258953 0.071405 8.7901e-05
readhsq(hsqfile="../result/imp1grm_CO_BE.hsq")
# H       SE     pvalue
# 1 0.415108 0.077425 7.2712e-09
# 2 0.268854 0.050146 7.2712e-09
readhsq(hsqfile="../result/imp1grm_CO_BE_recurrent1.hsq")
# H       SE   pvalue
# 1 0.358109 0.191492 0.021972
# 2 0.318422 0.170270 0.021972
readhsq(hsqfile="../result/imp1grm_CO_BE_recurrent0.hsq")
# H       SE   pvalue
# 1 0.242991 0.136421 0.039977
# 2 0.154112 0.086522 0.039977
readhsq(hsqfile="../result/imp1grm_CO_EA.hsq")
# H       SE     pvalue
# 1 0.461187 0.090805 6.6707e-08
# 2 0.191681 0.037741 6.6707e-08
readhsq(hsqfile="../result/imp1grm_CO_EA_recurrent1.hsq")
# H       SE     pvalue
# 1 0.972550 0.265861 0.00018817
# 2 0.492804 0.134715 0.00018817
readhsq(hsqfile="../result/imp1grm_CO_EA_recurrent0.hsq")
# H       SE  pvalue
# 1 0.179047 0.140658 0.12327
# 2 0.076885 0.060400 0.12327