#!/usr/bin/env Rscript
#USE GTEx mucosa data
# #get all the data
#generated by read_GTEx_EC_mucosa.R
#snp,snppos,rownames:1:1234_C_T


removehighcorr=function(dat=0,corcutoff=0.9)
{
  tmp <- cor(dat)
  tmp[upper.tri(tmp)] <- 0
  diag(tmp) <- 0
  datnew <- dat[,!apply(tmp,2,function(x) any(abs(x) > corcutoff))]
  return(datnew)
}

#predicted geneexp using crossvalidation
fitted_cv=function(Xsel,covariateall,Y,ncv=5)
{
  Xall=data.matrix(cbind(Xsel,covariateall))
  maxnumvar=floor(length(Y)*(1-1/ncv))
  if (ncol(Xall)>maxnumvar-1) #number of covariates is greater than sample size, select subset of covariates
  {
    lmfit1=lm(Y~Xall)
    lmcoeff1=summary(lmfit1)$coefficients
    rownames(lmcoeff1)=gsub("Xall","",rownames(lmcoeff1))
    lmleftsnp1=rownames(lmcoeff1)[rownames(lmcoeff1) %in% colnames(Xsel)] 
    idx1=match(lmleftsnp1,rownames(lmcoeff1))
    lmleftsnp1=lmleftsnp1[order(abs(lmcoeff1[idx1,1]),decreasing = T)]
    idx1=match(lmleftsnp1,colnames(Xsel))
    Xsel=Xsel[,idx1]
    Xsel=Xsel[,1:(maxnumvar-ncol(covariateall)-1)]
    
    Xall=data.matrix(cbind(Xsel,covariateall))
  }
  
  fitted1=rep(0,length(Y))
  set.seed(10000)
  permutidx=sample(1:length(Y))
  idxs=as.integer(seq(1,length(Y)+1,length.out = ncv+1)) #boundary points of cv segments
  
  for (ii in 1:ncv)
  {
    idx_predict=rep(F,length(Y))
    idx_predict[idxs[ii]:(idxs[ii+1]-1)]=T
    trainfm=lm(Y[permutidx[!idx_predict]]~Xall[permutidx[!idx_predict],])
    traincoeff=summary(trainfm)$coefficients
    rownames(traincoeff)=gsub("Xall[permutidx[!idx_predict], ]","",rownames(traincoeff),fixed = T)
    trainleftsnps=rownames(traincoeff)[rownames(traincoeff) %in% colnames(Xsel)]
    numvar=length(trainleftsnps)
    idx1=match(trainleftsnps,colnames(Xsel))
    Xsel1=Xsel[,idx1]
    if (numvar==1)
    {
      Xsel1=matrix(Xsel1,ncol=1)
      colnames(Xsel1)=trainleftsnps
    }
    fitted1[permutidx[idx_predict]]=rep(traincoeff[1,1],sum(idx_predict)) #intercept term
    idx1=match(trainleftsnps,rownames(traincoeff))
    if (numvar>0) #to add each selected snp term
    {
      for (j in 1:numvar)
      {
        fitted1[permutidx[idx_predict]]=fitted1[permutidx[idx_predict]]+Xsel1[permutidx[idx_predict],j]*traincoeff[idx1[j],1]
      }
    }
  }
  return(fitted1)
}

#new minimum rule,run 100 cvfit
compute_cor_arow=function(i,ncv=10,distcutoff=5e5)
{
  
  Y=unlist(phenotype[i,]) #geneexp
  r2=NA
  glmflag=0 #if glm selected variables
  tmp=distance(gr_snp,gr_pos[i])
  idx=which(tmp<distcutoff)
  tmp=rowSums(data.matrix(snp[idx,]))
  idx=idx[tmp!=0] #remove all 0 genotypes
  numvar=0 #number of snp selected by glmnet
  selectedsnps=NA
  selectedsnps_coeff=NA
  p_gender=NA
  p_age=NA
  tmp=quantile(Y,probs=c(0.15,0.85))
  #if (tmp[1]==tmp[2]) Y=Y+rnorm(length(Y),0,min(abs(Y))/1e6)
  if (tmp[1]!=tmp[2] & length(idx)>1)
  {
    #too many highly correlated SNPs are included for glmnet
    X1=t(snp[idx,])
    ucor <- matrix(0,ncol(X1),2)
    for (l in 1:ncol(X1)){
      ucov<- data.matrix(cbind(X1[,l],covariate))
      ufit <- lm(Y~ucov)
      ucor[l,1] <- summary(ufit)$coef[2,4]
      ucor[l,2] <- cor(Y,X1[,l])
    } 
    
    #hist(ucor[,1])
    pcor <- ucor[,1]
    ## I order the snps based on their correlation with gene expression, first delete those low correlation-with-gene SNPs
    X1 <- X1[,order(pcor,decreasing=T)]
    X <- removehighcorr(X1,0.8)
    if (class(X)=="numeric") #only 1 snp left
    {
      X=matrix(X,nrow=length(X),ncol=1)
    }
    #hist(ucor[colnames(X1)%in% colnames(X),1])
    #hist(ucor[colnames(X1)%in% colnames(X),2])
    #dim(X)
    
    Xall=data.matrix(cbind(X,covariate))
    
    covariateall=covariate
    
    penalty=rep(1,ncol(Xall))
    penalty[(ncol(X)+1):length(penalty)]=0 #force the covariates to be included in the model
    set.seed(i+10000)
    cvfit=tryCatch(
      {
        ### I change alpha to 0.5 for better variable selection when highly correlated features
        cv.glmnet(data.matrix(Xall),Y,nlambda=100,nfolds=10, penalty.factor=penalty,alpha=0.5)
      },
      error=function(e)
      {
        return(F)
      }
    )
    
    if (is.list(cvfit))
    {
      ## do 100 times cv.glmnet and take average for cverr
      ## the number of cvfit$lambda may be less than 100 sometimes even you specified 100
      cverr <- matrix(NA,length(cvfit$lambda),100)
      rownames(cverr)=cvfit$lambda
      for (l in 1:100) {
        set.seed(l+100)
        fit=tryCatch(
          {
            ### I change alpha to 0.5 for better variable selection when highly correlated features
            cv.glmnet(data.matrix(Xall),Y,nlambda=100,nfolds=10, penalty.factor=penalty,alpha=0.5)
          },
          error=function(e)
          {
            return(F)
          }
        )
        if (is.list(fit)) #Error in predmat[which, seq(nlami)] = preds : replacement has length zero
        {
          #fit <- cv.glmnet(data.matrix(Xall),Y,nlambda=100,nfolds=10, penalty.factor=penalty,alpha=0.5)
          alllambda=intersect(cvfit$lambda,fit$lambda)
          idx1=match(alllambda,cvfit$lambda)
          idx2=match(alllambda,fit$lambda)
          cverr[idx1,l] <- fit$cvm[idx2]
        }
      }
      merr <- apply(cverr,1,mean,na.rm=T)
      #plot(log(cvfit$lambda),merr)
      lambda.best <- cvfit$lambda[which.min(merr)]
      fit=glmnet(as.matrix(Xall),Y,nlambda = 100, penalty.factor=penalty,alpha=0.5)
      glmcoeff=as.matrix(coef(fit,s=lambda.best))
      #sum(rownames(glmcoeff)[glmcoeff[,1]!=0] %in% colnames(X))
      #glmcoeff[glmcoeff[,1]!=0 & rownames(glmcoeff) %in% colnames(X),1]
      glmleftsnp=rownames(glmcoeff)[rownames(glmcoeff) %in% colnames(X) & glmcoeff[,1]!=0] #snps left in glm model
      idx1=match(glmleftsnp,rownames(glmcoeff))
      glmleftsnp=glmleftsnp[order(abs(glmcoeff[idx1,1]),decreasing = T)] #order selected snp by effect size
      numvar=length(glmleftsnp)
      if (numvar>0)
      {
        idx1=match(glmleftsnp,colnames(X))
        Xsel=X[,idx1]
        if (numvar>1) #check if number of covariate is greater than sample size
        {
          nummaxvar=min(nrow(X)-ncol(covariateall)-1,numvar)
          numvar=nummaxvar
          Xsel=Xsel[,1:nummaxvar]
        }
        if (numvar==1) #keep Xsel as in matrix form
        {
          Xsel=matrix(Xsel,ncol=1)
          colnames(Xsel)=glmleftsnp
        }
        Xall1=data.matrix(cbind(Xsel,covariateall))
        #colnames(Xall1)[1:numvar]=glmleftsnp #deal with when only 1 snp is selected
        fit1=lm(Y~Xall1) # to remove snps with NA coefficient due to colinearity
        #summary(fit1)$r.squared
        lmcoeff=summary(fit1)$coefficients
        #align up coeff with Xsel
        rownames(lmcoeff)=gsub("Xall1","",rownames(lmcoeff))
        if (sum(rownames(lmcoeff)=="age")>0) p_age=lmcoeff[which(rownames(lmcoeff)=="age"),4]
        #p_disease=lmcoeff[which(rownames(lmcoeff)=="disease"),4]
        if (sum(rownames(lmcoeff)=="gender")>0) p_gender=lmcoeff[which(rownames(lmcoeff)=="gender"),4]
        lmleftsnp=rownames(lmcoeff)[rownames(lmcoeff) %in% colnames(Xsel)] 
        numvar=length(lmleftsnp)
        if (numvar>0)
        {
          glmflag=1
          idx1=match(lmleftsnp,rownames(lmcoeff))
          selectedsnps=paste0(rownames(lmcoeff)[idx1],collapse = "|")
          selectedsnps_coeff=paste0(lmcoeff[idx1,1],collapse = "|")
          idx1=match(lmleftsnp,colnames(Xsel))
          Xsel=Xsel[,idx1]
          if (numvar==1)
          {
            Xsel=matrix(Xsel,ncol=1)
            colnames(Xsel)=lmleftsnp
          }
          fitted=fitted_cv(Xsel,covariateall,Y,ncv=ncv)
          r2=cor(fitted,Y)^2
        }
      }
    }
  }
  return(list(r2=r2,glmflag=glmflag,numvar=numvar,numsnp=length(idx),selectedsnps=selectedsnps,selectedsnps_coeff=selectedsnps_coeff,
              p_age=p_age,p_gender=p_gender))
}

#load data
load("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/GTExV8mucosadatafor_prediction.RData")
library(glmnet)
library(GenomicRanges)
snppos$chr[snppos$chr==23]="X"
phenotypepos$chr[phenotypepos$chr==23]="X"
gr_allsnp=gr_snp=GRanges(seqnames = snppos$chr,ranges=IRanges(start=snppos$pos,width = 1)) #SNP
gr_allpos=gr_pos=GRanges(seqnames = phenotypepos$chr,ranges=IRanges(start=phenotypepos$s1,end=phenotypepos$s2)) #geneexp
allsnp=snp
allsnppos=snppos
allphenotype=phenotype
allphenotypepos=phenotypepos


#compute
Sys.time()
gene="FOXF1"
gene="THAP6"
test=compute_cor_arow(i=which(rownames(phenotype)==gene),ncv=10,distcutoff = 5e5)

test$numvar
Sys.time()
# #parallel
#salloc -t 6-1  -n 45 mpirun -n 1 R --interactive
#load("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/GTExmucosadatafor_prediction.RData")
library(Rmpi)
njobs=mpi.universe.size() - 1
print(njobs)
mpi.spawn.Rslaves(nslaves=njobs,needlog = F)
#mpi.remote.exec(load("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/GTExV8mucosadatafor_prediction.RData"))
#mpi.remote.exec(sum(is.na(snp)))
# mpi.bcast.Robj2slave(snppos)
# mpi.bcast.Robj2slave(phenotype)
# mpi.bcast.Robj2slave(phenotypepos)
# mpi.bcast.Robj2slave(copynumber)
# mpi.bcast.Robj2slave(covariate)
#mpi.bcast.Robj2slave(gr_snp)
#mpi.bcast.Robj2slave(gr_pos)
mpi.bcast.cmd(library(GenomicRanges))
mpi.bcast.cmd(library(glmnet))
mpi.bcast.Robj2slave(compute_cor_arow)
mpi.bcast.Robj2slave(fitted_cv)
mpi.bcast.Robj2slave(removehighcorr)
mpi.bcast.Robj2slave(covariate)

#run on each chr
mpi_compute_cor=function(ncv=10,distcutoff=5e5)
{
  res=NULL
  n=100
  for (chr in 1:22)
  {
    print(paste0(chr,"---"))
    idx=which(allphenotypepos$chr==chr)
    phenotype=allphenotype[idx,]
    phenotypepos=allphenotypepos[idx,]
    gr_pos=gr_allpos[idx]
    idx=which(allsnppos$chr==chr)
    snp=allsnp[idx,]
    gr_snp=gr_allsnp[idx]
    mpi.bcast.Robj2slave(gr_snp)
    mpi.bcast.Robj2slave(gr_pos)
    mpi.bcast.Robj2slave(phenotype)
    mpi.bcast.Robj2slave(snp)
    rows=1:nrow(phenotype)
    nchunks=ceiling(length(rows)/n)
    print(paste0("number of total:",nchunks))
    for (i in 1:nchunks)
    {
      cat(i,"..")
      if (i<nchunks)
      {
        seq=rows[((i-1)*n+1):(i*n)]
      }else
      {
        seq=rows[((i-1)*n+1):length(rows)]
      }
      tmp=mpi.parSapply(X=seq,FUN=compute_cor_arow,ncv=ncv,distcutoff=distcutoff,job.num=njobs)
      if (length(unlist(tmp)) %% 8 !=0) print (i)
      res1=matrix(unlist(tmp),ncol=8,byrow = T)
      rownames(res1)=rownames(phenotypepos)[seq]
      res=rbind(res,res1)
    }
    save(res,file="../result/tmp_res_min_GTExV8mucosa.RData")
  }
  colnames(res)=c("r2","glmflag","numselectedsnp","numtotalsnp","selectedsnps","selectedsnps_coeff","p_age","p_gender")
  res=as.data.frame(res)
  
  res[,5]=as.character(res[,5])
  res[,6]=as.character(res[,6])
  for (i in c(1:4,7:ncol(res))) res[,i]=as.numeric(as.character(res[,i]))
  return(res)
}

# mpi_compute_cor_old=function(rows=1:nrow(phenotypepos),ncv=10,distcutoff=1e6)
# {
#   res=NULL
#   n=1000
#   nchunks=ceiling(length(rows)/n)
#   print(paste0("number of total:",nchunks))
#   for (i in 1:nchunks)
#   {
#     cat(i,"..")
#     if (i<nchunks)
#     {
#       seq=rows[((i-1)*n+1):(i*n)]
#     }else
#     {
#       seq=rows[((i-1)*n+1):length(rows)]
#     }
#     tmp=mpi.parSapply(X=seq,FUN=compute_cor_arow,ncv=ncv,distcutoff=distcutoff,job.num=njobs)
#     if (length(unlist(tmp)) %% 8 !=0) print (i)
#     res1=matrix(unlist(tmp),ncol=8,byrow = T)
#     res=rbind(res,res1)
#   }
#   colnames(res)=c("r2","glmflag","numselectedsnp","numtotalsnp","selectedsnps","selectedsnps_coeff","p_age","p_gender")
#   res=as.data.frame(res)
#   rownames(res)=rownames(phenotypepos)
#   res[,5]=as.character(res[,5])
#   res[,6]=as.character(res[,6])
#   for (i in c(1:4,7:ncol(res))) res[,i]=as.numeric(as.character(res[,i]))
#   return(res)
# }

prefix="dist500K_GTEx_mucosa_May26"
outfolder=paste0("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/",prefix)
if (!dir.exists(outfolder)) dir.create(outfolder)
#rum mpi function and save the result
res_min=mpi_compute_cor(distcutoff = 5e5)
save(res_min,file=paste0(outfolder,"/preidiction_michigan_model.RData"))


#extract snps from the prediction models
extract_snp=function(dat=rbind(res_min))
{
  dat=dat[dat$glmflag==1,]
  snps=unique(unlist(strsplit(dat$selectedsnps,"|",fixed=T)))
  res=data.frame(snp=snps,chr=NA,pos=NA,stringsAsFactors = F)
  
  probenames=rownames(snppos)
  idx=match(res$snp,probenames)
  res$chr=snppos$chr[idx]
  res$pos=snppos$pos[idx]
  res=res[order(res[,2],res[,3]),]
  tmp=paste0(res[,2],"_",res[,3])
  idx=duplicated(tmp)
  res=res[!idx,]
  #create regions file used to extact genotype data
  for (i in 1:length(unique(res$chr)))
  {
    chr=unique(res$chr)[i]
    idx=which(res$chr==chr)
    chrres=paste0(chr,":",res$pos[idx],-res$pos[idx])
    chrres=paste0(chrres,collapse = "\n")
    filename=paste0(outfolder,"/prediction_snps_tabix_chr",chr,".txt")
    fileCon=file(filename)
    writeLines(chrres,fileCon)
    close(fileCon)
  }
  return(res)
}
#tmp=extract_snp()

extract_snp2=function(dat=rbind(res_min))
{
  dat=dat[dat$glmflag==1,]
  snps=unique(unlist(strsplit(dat$selectedsnps,"|",fixed=T)))
  res=data.frame(snp=snps,chr=NA,pos=NA,stringsAsFactors = F)
  
  probenames=rownames(snppos)
  idx=match(res$snp,probenames)
  res$chr=snppos$chr[idx]
  res$pos=snppos$pos[idx]
  idx=order(res$chr,res$pos)
  res=res[idx,]
  res=res[order(res[,2],res[,3]),]
  tmp=paste0(res[,2],"_",res[,3])
  idx=duplicated(tmp) #multi-allel
  res=res[!idx,]
  #create regions file used to extact genotype data
  for (i in 1:length(unique(res$chr)))
  {
    chr=unique(res$chr)[i]
    idx=which(res$chr==chr)
    tmp=ceiling(length(idx)/2)
    chrres=paste0(chr,":",res$pos[idx[1:tmp]],-res$pos[idx[1:tmp]])
    chrres=paste0(chrres,collapse = "\n")
    filename=paste0(outfolder,"/prediction_snps_tabix_chr",chr,"_1.txt")
    fileCon=file(filename)
    writeLines(chrres,fileCon)
    close(fileCon)
    
    chrres=paste0(chr,":",res$pos[idx[(tmp+1):length(idx)]],-res$pos[idx[(tmp+1):length(idx)]])
    chrres=paste0(chrres,collapse = "\n")
    filename=paste0(outfolder,"/prediction_snps_tabix_chr",chr,"_2.txt")
    fileCon=file(filename)
    writeLines(chrres,fileCon)
    close(fileCon)
  }
  return(res)
}

tmp0=extract_snp2()
cmd=paste0("./bca_extract_genotype2_V8.sh ",prefix)
system(cmd,wait = T)
#after run extract_genotype.sh, read the extracted genotype
library(data.table)
for (i in 1:22)
{
  cat(i,'..')
  tmp1=fread(paste0(outfolder,"/chr",i,"_select.bim"))
  tmp1=as.data.frame(tmp1)
  tmp=fread(paste0(outfolder,"/chr",i,"_select.raw"),header=T)
  tmp=as.data.frame(tmp)
  idx=duplicated(tmp$IID) #check this
  tmp=tmp[!idx,]
  rownames(tmp)=tmp$IID
  tmp=tmp[,7:ncol(tmp)]
  colnames(tmp)=paste0(tmp1[,1],":",tmp1[,4],"_",tmp1[,5],"_",tmp1[,6])
  if (i==1)
  {
    bcagenotype=tmp
    bcabim=tmp1
  }else
  {
    bcagenotype=cbind.data.frame(bcagenotype,tmp)
    bcabim=rbind(bcabim,tmp1)
  }
}
mycolnames=colnames(bcagenotype)
idx=duplicated(mycolnames)
bcagenotype=bcagenotype[,!idx]
bcagenotype=as.data.frame(t(bcagenotype))
bcabim=bcabim[!idx,]
save(bcagenotype,bcabim,file=paste0(outfolder,"/bca_extractgenotype.RData"))
# read_genotype=function(genotypefile="/fh/fast/dai_j/CancerGenomics/EAprogression/result/imputation_michigan/chr3_filter.raw",
#                        snpfile="/fh/fast/dai_j/CancerGenomics/EAprogression/result/imputation_michigan/chr3_filter.bim",
#                        snpname="3:196519878_C")
# {
#   genotype1=fread(genotypefile)
#   genotype1=as.data.frame(genotype1)
#   idx=which(colnames(genotype1)==snpname)
#   res=genotype1[,idx]
# }
# tcga_res=read_genotype()
# table(tcga_res)
# # 0  1  2 
# # 57 75 53
# bca_res=read_genotype(genotypefile = "/fh/fast/dai_j/BEACON/BEACON_GRANT/data/imputation/bca_1000g/chr3_select.raw",snpname ="3:196519878_T")
# table(bca_res)
# # 0    1    2 
# # 3173 4430 1600 
predict_geneexp=function(i=1,modeltable=res_min)
{
  idx=modeltable$glmflag==1
  #pick the lasso selected genes
  modeltable=modeltable[idx,]
  predicted_geneexp=data.frame(matrix(NA,nrow=1,ncol=2+ncol(bcagenotype_chunk)))
  rownames(predicted_geneexp)=rownames(modeltable)[i]
  colnames(predicted_geneexp)=c("n_totalsnp","n_avaisnp",colnames(bcagenotype_chunk))
  predicted_geneexp[,1]=modeltable$numselectedsnp[i]
  selectedsnps=unlist(strsplit(modeltable$selectedsnps[i],"|",fixed=T))
  selectedcoeff=as.numeric(unlist(strsplit(modeltable$selectedsnps_coeff[i],"|",fixed=T)))
  
  #if some imputed snps need to to flipped
  
  correctedsnps=NULL
  if (length(intersect(selectedsnps,rownames(bcagenotype_chunk)))<length(selectedsnps))
  {
    missingsnps=selectedsnps[!selectedsnps %in% rownames(bcagenotype_chunk)]
    for (j in 1:length(missingsnps))
    {
      tmp=unlist(strsplit(missingsnps[j],"_"))
      tmp1=paste0(tmp[c(1,3,2)],collapse = "_")  #change the order of allele
      idx=which(rownames(bcagenotype_chunk)==tmp1)
      if (length(idx)>0)
      {
        correctedsnps=c(correctedsnps,tmp1)
        idx1=which(selectedsnps==missingsnps[j])
        selectedsnps[idx1]=tmp1 #change the snp name to make it consistent with bca
      }
      
    }
  }
  idx=match(selectedsnps,rownames(bcagenotype_chunk))
  navaisnp=sum(!is.na(idx))
  if (navaisnp>0)
  {
    predicted_geneexp[,2]=navaisnp
    idx=selectedsnps %in% rownames(bcagenotype_chunk)
    selectedsnps=selectedsnps[idx]
    selectedcoeff=selectedcoeff[idx]
    idx1=match(selectedsnps,rownames(bcagenotype_chunk))
    availmat=bcagenotype_chunk[idx1,] #create a small matrix to avoid too large memory usage
    if (length(correctedsnps)>0)
    {
      idx2=match(correctedsnps,rownames(availmat))
      availmat[idx2,]=2-availmat[idx2,]
    }
    geneexp=as.matrix(t(availmat)) %*% selectedcoeff
    predicted_geneexp[,3:ncol(predicted_geneexp)]=geneexp
  }
  return(predicted_geneexp)
}

##NOTE: this step requires large nodes. It is not reliable to run on regular nodes (hang frequently)
##salloc -t 6-1 --mem-per-cpu 32G -n 21 --partition=largenode mpirun -n 1 R --interactive
if (!exists("phenotypepos")) load("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/GTExV8mucosadatafor_prediction.RData") #phenotypepos
load(paste0(outfolder,"/preidiction_michigan_model.RData")) #models
load(paste0(outfolder,"/bca_extractgenotype.RData")) #extracted genotype
library(Rmpi)
njobs=mpi.universe.size() - 1
print(njobs)
mpi.spawn.Rslaves(nslaves=njobs,needlog = F)
mpi.bcast.Robj2slave(res_min)
#mpi.bcast.Robj2slave(res_1se)
mpi.bcast.Robj2slave(outfolder)
#mpi.remote.exec(load(paste0(outfolder,"/bca_extractgenotype.RData")))
mpi.bcast.Robj2slave(bcabim)
mpi.bcast.Robj2slave(predict_geneexp)
#mpi.bcast.Robj2slave(snppos)
mpi_predict_geneexp=function(modeltable=res_min)
{
  modeltable=modeltable[modeltable$glmflag==1,]
  rows=1:nrow(modeltable)
  snps_bcagenotype=unlist(strsplit(rownames(bcagenotype),"_"))
  snps_bcagenotype=snps_bcagenotype[seq(1,length(snps_bcagenotype),3)]
  modeltable$cumsumsnp=cumsum(modeltable$numselectedsnp)
  nchunks=ceiling(sum(modeltable$numselectedsnp)/1500) #each chunk picks ~1500 snps
  joblables=cut(modeltable$cumsumsnp,nchunks)
  res=NULL
  # n=njobs
  # nchunks=ceiling(length(rows)/n)
  print(paste0("number of total:",nchunks))
  for (i in 1:nchunks)
  {
    if (i %% 5==0) cat(i,"..")
    #if (i %% 10==0) mpi.remote.exec(gc())
    seq=which(joblables==levels(joblables)[i])
    selectedsnps=unique(unlist(strsplit(modeltable$selectedsnps[seq],"|",fixed=T)))
    snps_selectedsnps=unlist(strsplit(selectedsnps,"_"))
    snps_selectedsnps=snps_selectedsnps[seq(1,length(snps_selectedsnps),3)]
    idx=snps_bcagenotype %in% snps_selectedsnps #extract genotypes for each chunk (use chr:pos)
    bcagenotype_chunk=bcagenotype[idx,]
    mpi.bcast.Robj2slave(bcagenotype_chunk)
    tmp=mpi.parSapply(X=seq,FUN=predict_geneexp,modeltable=modeltable,job.num=min(c(njobs,length(seq))))
    res1=as.data.frame(matrix(unlist(tmp),ncol=2+ncol(bcagenotype),byrow = T))
    if (length(unlist(tmp)) %% (2+ncol(bcagenotype)) !=0) stop(i)
    res=rbind(res,res1)
  }
  res=as.data.frame(res)
  rownames(res)=rownames(modeltable)
  colnames(res)=c("n_totalsnp","n_avaisnp",colnames(bcagenotype))
  return(res)
}
predict_min=mpi_predict_geneexp()
#predict_1se=mpi_predict_geneexp(modeltable=res_1se)
save(predict_min,file=paste0(outfolder,"/bca_predict_geneexp.RData"))
#after run mpi_predict_geneexp

#sequantial version, for some reason mpi is very slow
res=NULL
for (i in 1:nchunks)
{
  if (i %%5 ==0) cat(i,"..")
  seq=which(joblables==levels(joblables)[i])
  selectedsnps=unlist(strsplit(modeltable$selectedsnps[seq],"|",fixed=T))
  snps_selectedsnps=unlist(strsplit(selectedsnps,"_"))
  snps_selectedsnps=snps_selectedsnps[seq(1,length(snps_selectedsnps),3)]
  idx=snps_bcagenotype %in% snps_selectedsnps #extract genotypes for each chunk (use chr:pos)
  bcagenotype_chunk=bcagenotype[idx,]
  for (j in seq)
  {
    tmp=predict_geneexp(modeltable=modeltable,i=j)
    res1=as.data.frame(matrix(unlist(tmp),ncol=2+ncol(bcagenotype),byrow = T))
    if (length(unlist(tmp)) %% (2+ncol(bcagenotype)) !=0) stop(i)
    res=rbind(res,res1)
  }
}
res=as.data.frame(res)
rownames(res)=rownames(modeltable)
colnames(res)=c("n_totalsnp","n_avaisnp",colnames(bcagenotype))

#check predicted geneexp
tmp=rowMeans(predict_min[,3:ncol(predict_min)],na.rm=T)
sum(tmp==0,na.rm=T) #21
sum(is.na(tmp)) #6
which(tmp==0)
# LOC728788 ANKRD20A4      BCL8    BMS1P5  C1orf152     CBWD6   CXADRP2    FAM35B   FOXD4L6 
# 18       547      1156      1227      1741      2422      3516      4764      5175 
# GPRIN2 HIST1H2BL  HIST1H4E  HIST1H4K  KIAA0368 LOC645166 LOC646214 LOC648691 LOC654342 
# 5745      6084      6099      6102      6905      7736      7741      7751      7760 
# PGM5P2    SPDYE1     SYT15 
# 10062     12874     13234 

#association
library(readxl)
#if (!exists("sampletable"))
#{
   # sampletable=read.table("/fh/fast/dai_j/BEACON/BEACON_GRANT/data//bc_AT_JD1.csv",header=T,sep=",",stringsAsFactors=F)
  # sampletable$site[sampletable$site=="NA"]=NA
  sampletable=read_excel("/fh/fast/dai_j/BEACON/BEACON_GRANT/data/PLINKinputCombo_bca_07Feb2018.xls",1)
  geneexpsamplenames=strsplit(colnames(predict_min)[3:ncol(predict_min)],"_") #use localid
  geneexpsamplenames=sapply(1:length(geneexpsamplenames),function(x){
    tmp=geneexpsamplenames[[x]]
    paste0(tmp[2:length(tmp)],collapse = "_")
  })
#}

# sampletablenew= read_excel("/fh/fast/dai_j/BEACON/BEACON_GRANT/data/PLINKinputCombo_bca_07Feb2018.xls",1)
# colnames(sampletablenew)[!colnames(sampletablenew) %in% colnames(sampletable)]
# colnames(sampletable)[!colnames(sampletable) %in% colnames(sampletablenew)]
# colnames(sampletable)[colnames(sampletable) %in% colnames(sampletablenew)]
# all(sampletable$localid %in% sampletablenew$localid)
# idx=match(sampletable$localid,sampletablenew$localid)
# all(sampletable$sex==sampletablenew$sex[idx])
# all(sampletable$age==sampletablenew$age[idx])
# all(sampletable$site==sampletablenew$site[idx],na.rm=T)
# all(sampletable$phenoBE_bc==sampletablenew$phenoBE_bca[idx])
# all(sampletable$phenoEA_bc==sampletablenew$phenoEA_bca[idx])
# all(sampletable$phenoEABE_bc==sampletablenew$phenoEABE_bca[idx])
# all(sampletable$cig_smk_ever==sampletablenew$cig_smk_ever[idx])
# all(sampletable$recurrent_HB_RF==sampletablenew$recurrent_HB_RF[idx])
# sum(sampletable$bmi_recent_healthy==sampletablenew$bmi_recent_healthy[idx])
# plot(sampletable$bmi_recent_healthy,sampletablenew$bmi_recent_healthy[idx])
# max(abs(sampletable$bmi_recent_healthy-sampletablenew$bmi_recent_healthy[idx]))
# plot(sampletable$bmi_current,sampletable$bmi_recent_healthy)
# plot(1:nrow(sampletable),sampletable$bmi_current)
# plot(1:nrow(sampletable),sampletable$bmi_recent_healthy)
for (i in 1:ncol(sampletable))
{
  idx=which(sampletable[,i]==-9)
  sampletable[idx,i]=NA
}
# readeigenstrat=function(eigfile="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/bca_filtered_10Jan2019.pca",
#                         eigsampfile="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/bca_filtered_10Jan2019.pedind",
#                         thesamples=geneexpsamplenames,nskip=16,opt=1)
# {
#   tmp=read.table(eigfile,skip=nskip,stringsAsFactors = F)
#   if (opt==1)
#   {
#     eigsamples=read.table(eigsampfile,stringsAsFactors = F)
#     eigsamples=eigsamples$V2
#     idx=match(thesamples,eigsamples)
#     tmp1=tmp[idx,]
#     tmp1=as.data.frame(t(tmp1))
#     colnames(tmp1)=thesamples
#   }else #don't need to change sample names
#   {
#     tmp1=as.data.frame(t(tmp))
#     colnames(tmp1)=thesamples
#   }
#   rownames(tmp1)=paste0("pc",1:nrow(tmp1))
#   return(tmp1)
# }
# 
# eigenstratmatrix=readeigenstrat()

pvalue_arow=function(i,idx1,idx2)
{
  y=c(rep(1,length(idx1)),rep(0,length(idx2)))
  x=as.numeric(predict_bcageneexp[i,c(idx1,idx2)])
  BE_p=NA
  if (sum(is.na(x))<0.6*length(x))
  {
    fm=glm(y~x+age+sex+pc1+pc2+pc3+pc4,data=covariates,family=binomial)
    #fm=glm(y~x+age+sex+site1+pc1+pc2+pc3,data=covariates,family=binomial) #including site
    tmp=summary(fm)$coefficients
    if ("x" %in% rownames(tmp))
    {
      BE_p=summary(fm)$coefficients[2,4]
    }
  }
  return(BE_p)
}


qqplot=function(pvalue=NULL,main="",xlim=NULL,ylim=NULL)
{
  n=length(pvalue)
  plot(-log((1:n)/n,base=10),-log(pvalue[order(pvalue)],base=10),xlab="Expected p-value (log10)",
       ylab="Observed p-value (log10)",main=main,xlim=xlim,ylim=ylim,cex.lab=1.3,cex.axis=1.3)
  abline(0,1,lty=2)
}

load(paste0(outfolder,"/bca_predict_geneexp.RData"))
# recodesite=function(covariates1=covariates,idx1,idx2)
# {
#   caseidx=1:length(idx1)
#   coidx=(length(idx1)+1):nrow(covariates1)
#   covariates1$site1=NA
#   for (mysite in unique(covariates1$site))
#   {
#     idx=which(covariates1$site==mysite)
#     if (sum(idx %in% caseidx)>0 && sum(idx %in% coidx)>0)
#     {
#       covariates1$site1[idx]=mysite
#     }else
#     {
#       covariates1$site1[idx]="other"
#     }
#   }
#   return(covariates1)
# }
mpi_association=function(predict_bcageneexp=predict_min)
{
  predict_bcageneexp=predict_bcageneexp[,3:ncol(predict_bcageneexp)]
  geneexpsamplenames=colnames(predict_bcageneexp)
  geneexpsamplenames=strsplit(geneexpsamplenames,"_")
  geneexpsamplenames=sapply(1:length(geneexpsamplenames),function(x){
    tmp=geneexpsamplenames[[x]]
    paste0(tmp[2:length(tmp)],collapse = "_")
  })
  colnames(predict_bcageneexp)=geneexpsamplenames
  geneexpsamplenames=intersect(sampletable$localid,geneexpsamplenames)
  idx=match(geneexpsamplenames,colnames(predict_bcageneexp))
  all_predict_bcageneexp=predict_bcageneexp=predict_bcageneexp[,idx]
  idx=match(geneexpsamplenames,sampletable$localid)
  sampletable1=sampletable[idx,]
  for (i in 1:ncol(sampletable1)) sampletable1[,i][sampletable1[,i]==-9]=NA
  # idx=match(geneexpsamplenames,colnames(eigenstratmatrix))
  # eigenstratmatrix=eigenstratmatrix[,idx]
  # sampletable1=cbind.data.frame(sampletable1,t(eigenstratmatrix[1:4,]))
  #BE
  mpi.bcast.Robj2slave(pvalue_arow)
  #mpi.bcast.Robj2slave(recodesite)
  #mpi.bcast.Robj2slave(predict_bcageneexp)
  mpi.bcast.Robj2slave(sampletable1)
  idx1=which(sampletable1$phenoBE_bca==2) #case
  length(idx1) #3288
  idx2=which(sampletable1$phenoBE_bca==1)
  length(idx2) #3195
  covariates=data.frame(age=sampletable1$age[c(idx1,idx2)],sex=sampletable1$sex[c(idx1,idx2)],site=sampletable1$site[c(idx1,idx2)],
                        bmi=sampletable1$bmi_recent_healthy[c(idx1,idx2)],cig_smk=sampletable1$cig_smk_ever[c(idx1,idx2)],cig_pack_yrs=sampletable1$cig_pack_yrs[c(idx1,idx2)],
                        pc1=sampletable1$ev1_bca[c(idx1,idx2)],pc2=sampletable1$ev2_bca[c(idx1,idx2)],pc3=sampletable1$ev3_bca[c(idx1,idx2)],pc4=sampletable1$ev4_bca[c(idx1,idx2)],
                        stringsAsFactors = F)
  #covariates=recodesite(covariates1=covariates,idx1=idx1,idx2=idx2)
  mpi.bcast.Robj2slave(idx1)
  mpi.bcast.Robj2slave(idx2)
  mpi.bcast.Robj2slave(covariates)
  n=1000 #1000 snps one time
  rows=1:nrow(all_predict_bcageneexp)
  nchunks=ceiling(nrow(all_predict_bcageneexp)/n)
  print(paste0("BE:number of total:",nchunks))
  BE_p=NULL
  for (i in 1:nchunks)
  {
    cat(i,"..")
    if (i<nchunks)
    {
      seq=rows[((i-1)*n+1):(i*n)]
    }else
    {
      seq=rows[((i-1)*n+1):length(rows)]
    }
    predict_bcageneexp=all_predict_bcageneexp[seq,]
    mpi.bcast.Robj2slave(predict_bcageneexp)
    seq=1:nrow(predict_bcageneexp)
    tmp=mpi.parSapply(X=seq,FUN=pvalue_arow,idx1=idx1,idx2=idx2,job.num=njobs)
    BE_p=c(BE_p,tmp)
  }
  
  
  #EA
  idx1=which(sampletable1$phenoEA_bca==2) #case
  length(idx1) #2514
  idx2=which(sampletable1$phenoEA_bca==1)
  length(idx2) #3198
  covariates=data.frame(age=sampletable1$age[c(idx1,idx2)],sex=sampletable1$sex[c(idx1,idx2)],site=sampletable1$site[c(idx1,idx2)],
                        bmi=sampletable1$bmi_recent_healthy[c(idx1,idx2)],cig_smk=sampletable1$cig_smk_ever[c(idx1,idx2)],cig_pack_yrs=sampletable1$cig_pack_yrs[c(idx1,idx2)],
                        pc1=sampletable1$ev1_bca[c(idx1,idx2)],pc2=sampletable1$ev2_bca[c(idx1,idx2)],pc3=sampletable1$ev3_bca[c(idx1,idx2)],pc4=sampletable1$ev4_bca[c(idx1,idx2)],
                        stringsAsFactors = F)
  #covariates=recodesite(covariates1=covariates,idx1=idx1,idx2=idx2)
  mpi.bcast.Robj2slave(idx1)
  mpi.bcast.Robj2slave(idx2)
  mpi.bcast.Robj2slave(covariates)
  print(paste0("EA:number of total:",nchunks))
  EA_p=NULL
  for (i in 1:nchunks)
  {
    cat(i,"..")
    if (i<nchunks)
    {
      seq=rows[((i-1)*n+1):(i*n)]
    }else
    {
      seq=rows[((i-1)*n+1):length(rows)]
    }
    predict_bcageneexp=all_predict_bcageneexp[seq,]
    mpi.bcast.Robj2slave(predict_bcageneexp)
    seq=1:nrow(predict_bcageneexp)
    tmp=mpi.parSapply(X=seq,FUN=pvalue_arow,idx1=idx1,idx2=idx2,job.num=njobs)
    EA_p=c(EA_p,tmp)
  }
  
  #EA vs # BE
  idx1=which(sampletable1$phenoEA_bca==2) # EAcase
  length(idx1) #2514
  idx2=which(sampletable1$phenoBE_bca==2) #BEcase
  length(idx2) #3288
  length(intersect(idx1,idx2))
  covariates=data.frame(age=sampletable1$age[c(idx1,idx2)],sex=sampletable1$sex[c(idx1,idx2)],site=sampletable1$site[c(idx1,idx2)],
                        bmi=sampletable1$bmi_recent_healthy[c(idx1,idx2)],cig_smk=sampletable1$cig_smk_ever[c(idx1,idx2)],cig_pack_yrs=sampletable1$cig_pack_yrs[c(idx1,idx2)],
                        pc1=sampletable1$ev1_bca[c(idx1,idx2)],pc2=sampletable1$ev2_bca[c(idx1,idx2)],pc3=sampletable1$ev3_bca[c(idx1,idx2)],pc4=sampletable1$ev4_bca[c(idx1,idx2)],
                        stringsAsFactors = F)
  #covariates=recodesite(covariates1=covariates,idx1=idx1,idx2=idx2)
  mpi.bcast.Robj2slave(idx1)
  mpi.bcast.Robj2slave(idx2)
  mpi.bcast.Robj2slave(covariates)
  print(paste0("BEA:number of total:",nchunks))
  BEA_p=NULL
  for (i in 1:nchunks)
  {
    cat(i,"..")
    if (i<nchunks)
    {
      seq=rows[((i-1)*n+1):(i*n)]
    }else
    {
      seq=rows[((i-1)*n+1):length(rows)]
    }
    predict_bcageneexp=all_predict_bcageneexp[seq,]
    mpi.bcast.Robj2slave(predict_bcageneexp)
    seq=1:nrow(predict_bcageneexp)
    tmp=mpi.parSapply(X=seq,FUN=pvalue_arow,idx1=idx1,idx2=idx2,job.num=njobs)
    BEA_p=c(BEA_p,tmp)
  }
  
  #EA and BE vs control
  idx1=which(sampletable1$phenoBE_bca==2 | sampletable1$phenoEA_bca==2) #BE or EAcase
  length(idx1) #5802
  idx2=which(sampletable1$phenoBE_bca==1 |sampletable1$phenoEA_bca==1) #Control
  length(idx2) #3198
  length(intersect(idx1,idx2))
  covariates=data.frame(age=sampletable1$age[c(idx1,idx2)],sex=sampletable1$sex[c(idx1,idx2)],site=sampletable1$site[c(idx1,idx2)],
                        bmi=sampletable1$bmi_recent_healthy[c(idx1,idx2)],cig_smk=sampletable1$cig_smk_ever[c(idx1,idx2)],cig_pack_yrs=sampletable1$cig_pack_yrs[c(idx1,idx2)],
                        pc1=sampletable1$ev1_bca[c(idx1,idx2)],pc2=sampletable1$ev2_bca[c(idx1,idx2)],pc3=sampletable1$ev3_bca[c(idx1,idx2)],pc4=sampletable1$ev4_bca[c(idx1,idx2)],
                        stringsAsFactors = F)
  #covariates=recodesite(covariates1=covariates,idx1=idx1,idx2=idx2)
  mpi.bcast.Robj2slave(idx1)
  mpi.bcast.Robj2slave(idx2)
  mpi.bcast.Robj2slave(covariates)
  print(paste0("BEEA:number of total:",nchunks))
  BEEA_p=NULL
  for (i in 1:nchunks)
  {
    cat(i,"..")
    if (i<nchunks)
    {
      seq=rows[((i-1)*n+1):(i*n)]
    }else
    {
      seq=rows[((i-1)*n+1):length(rows)]
    }
    predict_bcageneexp=all_predict_bcageneexp[seq,]
    mpi.bcast.Robj2slave(predict_bcageneexp)
    seq=1:nrow(predict_bcageneexp)
    tmp=mpi.parSapply(X=seq,FUN=pvalue_arow,idx1=idx1,idx2=idx2,job.num=njobs)
    BEEA_p=c(BEEA_p,tmp)
  }
  
  res=data.frame(BE_p=BE_p,BE_fdr=p.adjust(BE_p),EA_p=EA_p,EA_fdr=p.adjust(EA_p),BEA_p=BEA_p,BEA_fdr=p.adjust(BEA_p),BEEA_p=BEEA_p,BEEA_fdr=p.adjust(BEEA_p))
  rownames(res)=rownames(all_predict_bcageneexp)
  return(res)
}
assoc_min=mpi_association()
#assoc_1se=mpi_association(predict_bcageneexp = predict_1se)
save(assoc_min,file=paste0(outfolder,"/bca_assoc.RData"))
load(paste0(outfolder,"/bca_assoc.RData"))
quantile(assoc_min[,1],na.rm=T)
quantile(assoc_min[,3],na.rm=T)
quantile(assoc_min[,5],na.rm=T)
quantile(assoc_min[,7],na.rm=T)
# quantile(assoc_1se[,1],na.rm=T)
# quantile(assoc_1se[,3],na.rm=T)
# quantile(assoc_1se[,5],na.rm=T)
# quantile(assoc_1se[,7],na.rm=T)


load("/fh/fast/stanford_j/Xiaoyu/QTL/data/GTEx/gtex_ge_anno.RData")
prefix="dist500K_GTEx"
proteingenes=unique(gtex_ge_anno$Symbol[gtex_ge_anno$gene_type=="protein_coding"])
qqplot(assoc_min[,1])
idx1=which(rownames(assoc_min) %in% proteingenes)
qqplot(assoc_min[idx1,1])
which.min(assoc_min[idx1,1])
assoc_min[idx1[4307],]
#           BE_p    BE_fdr       EA_p EA_fdr      BEA_p BEA_fdr       BEEA_p BEEA_fdr
# MYOM2 1.401227e-05 0.3694616 0.02769562      1 0.05926998       1 0.0001009403        1
tmp=p.adjust(assoc_min[idx1,1])
tmp[4307] #fdr=0.15 (based on proteins)
res_min$r2[which(rownames(res_min)=="MYOM2")] #0.25
res_min$numselectedsnp[which(rownames(res_min)=="MYOM2")] #124
idx2=which(rownames(assoc_min) %in% rownames(res_min)[res_min$r2>0.05])
idx3=intersect(idx1,idx2)
qqplot(assoc_min[idx3,1])
which.min(assoc_min[idx3,1])
assoc_min[idx3[2492],]
tmp=p.adjust(assoc_min[idx3,1])
tmp[2492] #fdr=0.087


qqplot(assoc_min[,3])
idx1=which(rownames(assoc_min) %in% proteingenes)
qqplot(assoc_min[idx1,3])
which.min(assoc_min[idx1,3]) #9035
assoc_min[idx1[9035],]
tmp=p.adjust(assoc_min[idx1,3])
tmp[9035] #fdr=0.98 (based on proteins)
qqplot(assoc_min[idx3,3])

qqplot(assoc_min[,5])
idx1=which(rownames(assoc_min) %in% proteingenes)
qqplot(assoc_min[idx1,5])
which.min(assoc_min[idx1,5]) #5054
assoc_min[idx1[5054],]
#           BE_p BE_fdr       EA_p EA_fdr        BEA_p    BEA_fdr    BEEA_p BEEA_fdr
# DPP7 0.01628854      1 0.02823387      1 3.020026e-06 0.07962901 0.6230114        1
#beta BE_p -0.175732, EA_p 0.173902
tmp=p.adjust(assoc_min[idx1,5])
tmp[5054] #fdr=0.03 (based on proteins)
res_min$r2[which(rownames(res_min)=="DPP7")] #0.11
res_min$numselectedsnp[which(rownames(res_min)=="DPP7")] #55
qqplot(assoc_min[idx3,5])
which.min(assoc_min[idx3,5]) #2946
assoc_min[idx3[2946],]
tmp=p.adjust(assoc_min[idx3,5])
names(tmp)=rownames(assoc_min)[idx3]
tmp[2946] #fdr=0.98 (based on proteins)

qqplot(assoc_min[,7])
idx1=which(rownames(assoc_min) %in% proteingenes)
qqplot(assoc_min[idx1,7])
which.min(assoc_min[idx1,7]) #9035
assoc_min[idx1[9035],]
#                 BE_p BE_fdr         EA_p EA_fdr     BEA_p BEA_fdr       BEEA_p  BEEA_fdr
# KIAA1683 9.736606e-05      1 9.410342e-05      1 0.9497299       1 1.453249e-05 0.3831926
tmp=p.adjust(assoc_min[idx1,7])
tmp[9035] #fdr=0.15 (based on proteins)
res_min$r2[which(rownames(res_min)=="KIAA1683")] #0.19
res_min$numselectedsnp[which(rownames(res_min)=="KIAA1683")] #119
qqplot(assoc_min[idx3,7])
which.min(assoc_min[idx3,7]) #5330
assoc_min[idx3[5330],]
tmp=p.adjust(assoc_min[idx3,7])
names(tmp)=rownames(assoc_min)[idx3]
tmp[5330] #fdr=0.09

idx1=which(sampletable1$phenoBE_bc==2)
idx2=which(sampletable1$phenoEA_bc==2)
idx3=which(sampletable1$phenoBE_bc==1)
which(rownames(predict_bcageneexp)=="DPP7") #9927
boxplot(unlist(predict_bcageneexp[9927,c(idx3,idx1,idx2)])~c(rep("CO",length(idx3)),rep("BE",length(idx1)),rep("EA",length(idx2))))
which(rownames(res_min)=="DPP7") #17859
selectedsnps=unlist(strsplit(res_min$selectedsnps[17859],"|",fixed=T))
selectedsnps[19]="9:140356374_C_A"
idx=match(selectedsnps,rownames(bcagenotype))
dat=t(bcagenotype[idx,])
cordat=cor(dat)
library(gplots)
heatmap.2(as.matrix(cordat), col=redgreen(75), 
          density.info="none", trace="none", dendrogram=c("row"), 
          symm=T,symkey=T,symbreaks=T, scale="none")
selectedsnps=unlist(strsplit(res_min$selectedsnps[17859],"|",fixed=T))
idx=match(selectedsnps,rownames(snp))
dat1=t(snp[idx,])
cordat1=cor(dat1)
heatmap.2(as.matrix(cordat1), col=redgreen(75), 
          density.info="none", trace="none", dendrogram=c("row"), 
          symm=F,symkey=T,symbreaks=T, scale="none")


qqplot(assoc_1se[,1])
idx1=which(rownames(assoc_1se) %in% proteingenes)
qqplot(assoc_1se[idx1,1])

qqplot(assoc_1se[,3])
idx1=which(rownames(assoc_1se) %in% proteingenes)
qqplot(assoc_1se[idx1,3])
which.min(assoc_1se[,3]) #7333
assoc_1se[idx1[7333],]
tmp=p.adjust(assoc_1se[,3])
tmp[7333] #fdr=0.27 (based on non-proteins)

qqplot(assoc_1se[,5])
idx1=which(rownames(assoc_1se) %in% proteingenes)
qqplot(assoc_1se[idx1,5])


qqplot(assoc_1se[,7])
idx1=which(rownames(assoc_1se) %in% proteingenes)
qqplot(assoc_1se[idx1,7])
which.min(assoc_1se[,7]) #6393
assoc_1se[6393,]
#           BE_p    BE_fdr         EA_p EA_fdr     BEA_p BEA_fdr       BEEA_p  BEEA_fdr
# FAM85B 9.91424e-05 0.7454517 0.0008693366      1 0.7316979       1 3.385742e-05 0.2545739
#                 BE_p BE_fdr         EA_p EA_fdr     BEA_p BEA_fdr       BEEA_p  BEEA_fdr
# KIAA1683 9.736606e-05      1 9.410342e-05      1 0.9497299       1 1.453249e-05 0.3831926
tmp=p.adjust(assoc_1se[idx1,7])
tmp[9035] #fdr=0.15 (based on proteins)

outfolder="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/dist500K_GTEx"
load(paste0(outfolder,"/preidiction_michigan_model.RData"))
gtex_res_min=res_min
gtex_res_1se=res_1se
load(paste0(outfolder,"/bca_assoc.RData"))
gtex_assoc_min=assoc_min
gtex_assoc_1se=assoc_1se
load(paste0(outfolder,"/bca_predict_geneexp.RData"))
gtex_predict_min=predict_min
gtex_predict_1se=predict_1se
outfolder="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/dist500K_EAC2"
load(paste0(outfolder,"/preidiction_michigan_model.RData"))
tcga_res_min=res_min
tcga_res_1se=res_1se
load(paste0(outfolder,"/bca_assoc.RData"))
tcga_assoc_min=assoc_min
tcga_assoc_1se=assoc_1se
load(paste0(outfolder,"/bca_predict_geneexp.RData"))
tcga_predict_min=predict_min
tcga_predict_1se=predict_1se

comgene=intersect(rownames(gtex_predict_min),rownames(tcga_predict_min))
idx=match(comgene,rownames(tcga_predict_min))
tcga_predict_min1=tcga_predict_min[idx,3:ncol(tcga_predict_min)]
idx=match(comgene,rownames(gtex_predict_min))
gtex_predict_min1=gtex_predict_min[idx,3:ncol(gtex_predict_min)]
#cor_predict_min=cor(t(gtex_predict_min1),t(tcga_predict_min1))
cor_diag_min=rep(NA,nrow(gtex_predict_min1))
for (i in 1:nrow(gtex_predict_min1)) 
{
  cor_diag_min[i]=cor(tcga_predict_min1[,i],gtex_predict_min1[,i])
}  
hist(cor_diag_min,main="",xlab="correlation")
# heatmap.2(as.matrix(cor_predict_min), col=redgreen(75), 
#           density.info="none", trace="none", dendrogram=c("row"), 
#           symm=T,symkey=T,symbreaks=T, scale="none")


#check RBBP5
load("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/dist500K_EAC/bca_predict_geneexp.RData")
predict_bcageneexp=predict_1se
predict_bcageneexp=predict_bcageneexp[,3:ncol(predict_bcageneexp)]
geneexpsamplenames=colnames(predict_bcageneexp)
geneexpsamplenames=strsplit(geneexpsamplenames,"_")
geneexpsamplenames=sapply(1:length(geneexpsamplenames),function(x){
  tmp=geneexpsamplenames[[x]]
  paste0(tmp[2:length(tmp)],collapse = "_")
})
colnames(predict_bcageneexp)=geneexpsamplenames
geneexpsamplenames=intersect(sampletable$localid,geneexpsamplenames)
idx=match(geneexpsamplenames,colnames(predict_bcageneexp))
predict_bcageneexp=predict_bcageneexp[,idx]
idx=match(geneexpsamplenames,sampletable$localid)
sampletable1=sampletable[idx,]
for (i in 1:ncol(sampletable1)) sampletable1[,i][sampletable1[,i]==-9]=NA
idx=match(geneexpsamplenames,colnames(eigenstratmatrix))
eigenstratmatrix=eigenstratmatrix[,idx]
sampletable1=cbind.data.frame(sampletable1,t(eigenstratmatrix[1:4,]))
#gene="RBBP5"
idx1=which(sampletable1$phenoEA_bc==2) # EAcase
length(idx1) #2514
idx2=which(sampletable1$phenoBE_bc==2) #BEcase
length(idx2) #3288knowndat2=knowndat2[idx,
covariates=data.frame(age=sampletable1$age[c(idx1,idx2)],sex=sampletable1$sex[c(idx1,idx2)],site=sampletable1$site[c(idx1,idx2)],
                      bmi=sampletable1$bmi_recent_healthy[c(idx1,idx2)],cig_smk=sampletable1$cig_smk_ever[c(idx1,idx2)],cig_pack_yrs=sampletable1$cig_pack_yrs[c(idx1,idx2)],
                      pc1=sampletable1$pc1[c(idx1,idx2)],pc2=sampletable1$pc2[c(idx1,idx2)],pc3=sampletable1$pc3[c(idx1,idx2)],
                      stringsAsFactors = F)
#covariates=recodesite(covariates1=covariates,idx1=idx1,idx2=idx2)
idxgene=which(rownames(predict_bcageneexp)==gene) #4642
pvalue_arow_sites=function(i,idx1,idx2)
{
  caseidx=1:length(idx1)
  coidx=(length(idx1)+1):nrow(covariates)
  y=c(rep(1,length(idx1)),rep(0,length(idx2)))
  x=as.numeric(predict_bcageneexp[i,c(idx1,idx2)])
  n=length(unique(covariates$site1))
  res=data.frame(matrix(NA,nrow=n,ncol=4))
  colnames(res)=c("p","coeff","ncase","nco")
  rownames(res)=unique(covariates$site1)
  for (i in 1:n)
  {
    idx=which(covariates$site1==unique(covariates$site1)[i])
    res$ncase[i]=sum(idx %in% caseidx)
    res$nco[i]=sum(idx %in% coidx)
    fm=glm(y[idx]~x[idx]+age+sex+pc1+pc2+pc3,data=covariates[idx,],family=binomial)
    #fm=glm(y~x+age+sex+site1+pc1+pc2+pc3,data=covariates,family=binomial) #including site
    tmp=summary(fm)$coefficients
    if ("x[idx]" %in% rownames(tmp))
    {
      res$p[i]=summary(fm)$coefficients[2,4]
      res$coeff[i]=summary(fm)$coefficients[2,1]
    }
  }
  return(res)
}

#check DDX49
gene="DDX49"
load("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/dist500K_EAC/bca_predict_geneexp.RData")
predict_bcageneexp=predict_1se
predict_bcageneexp=predict_bcageneexp[,3:ncol(predict_bcageneexp)]
geneexpsamplenames=colnames(predict_bcageneexp)
geneexpsamplenames=strsplit(geneexpsamplenames,"_")
geneexpsamplenames=sapply(1:length(geneexpsamplenames),function(x){
  tmp=geneexpsamplenames[[x]]
  paste0(tmp[2:length(tmp)],collapse = "_")
})
colnames(predict_bcageneexp)=geneexpsamplenames
geneexpsamplenames=intersect(sampletable$localid,geneexpsamplenames)
idx=match(geneexpsamplenames,colnames(predict_bcageneexp))
predict_bcageneexp=predict_bcageneexp[,idx]
idx=match(geneexpsamplenames,sampletable$localid)
sampletable1=sampletable[idx,]
for (i in 1:ncol(sampletable1)) sampletable1[,i][sampletable1[,i]==-9]=NA
idx=match(geneexpsamplenames,colnames(eigenstratmatrix))
eigenstratmatrix=eigenstratmatrix[,idx]
sampletable1=cbind.data.frame(sampletable1,t(eigenstratmatrix[1:4,]))
idx1=which(sampletable1$phenoBE_bc==2) #case
length(idx1) #3288
idx2=which(sampletable1$phenoBE_bc==1)
length(idx2) #2176
idx1=which(sampletable1$phenoBE_bc==2 | sampletable1$phenoEA_bc==2) #BE or EAcase
length(idx1) #5802
covariates=data.frame(age=sampletable1$age[c(idx1,idx2)],sex=sampletable1$sex[c(idx1,idx2)],site=sampletable1$site[c(idx1,idx2)],
                      bmi=sampletable1$bmi_recent_healthy[c(idx1,idx2)],cig_smk=sampletable1$cig_smk_ever[c(idx1,idx2)],cig_pack_yrs=sampletable1$cig_pack_yrs[c(idx1,idx2)],
                      pc1=sampletable1$pc1[c(idx1,idx2)],pc2=sampletable1$pc2[c(idx1,idx2)],pc3=sampletable1$pc3[c(idx1,idx2)],
                      stringsAsFactors = F)
covariates=recodesite(covariates1=covariates,idx1=idx1,idx2=idx2)
idxgene=which(rownames(predict_bcageneexp)==gene) #2182
DDX49res=pvalue_arow_sites(i=idxgene,idx1,idx2)

idx=which(res_1se_500k$r2>0.05)
idx=which(rownames(assoc_1se_500K) %in% rownames(res_1se_500k)[idx])
qqplot(assoc_1se_500K$BE_p[idx])
min(p.adjust(assoc_1se_500K$BE_p[idx]),na.rm=T)
rownames(assoc_1se_500K)[idx][which.min(p.adjust(assoc_1se_500K$BE_p[idx]))]
qqplot(assoc_1se_500K$EA_p[idx])
min(p.adjust(assoc_1se_500K$EA_p[idx]),na.rm=T)
rownames(assoc_1se_500K)[idx][which.min(p.adjust(assoc_1se_500K$EA_p[idx]))]
qqplot(assoc_1se_500K$BEA_p[idx])
min(p.adjust(assoc_1se_500K$BEA_p[idx]),na.rm=T)
rownames(assoc_1se_500K)[idx][which.min(p.adjust(assoc_1se_500K$BEA_p[idx]))]

idx=which(res_1se_1M$r2>0.05)
idx=which(rownames(assoc_1se_1M) %in% rownames(res_1se_1M)[idx])
qqplot(assoc_1se_1M$BE_p[idx])
min(p.adjust(assoc_1se_1M$BE_p[idx]),na.rm=T)
rownames(assoc_1se_1M)[idx][which.min(p.adjust(assoc_1se_1M$BE_p[idx]))]
qqplot(assoc_1se_1M$EA_p[idx])
min(p.adjust(assoc_1se_1M$EA_p[idx]),na.rm=T)
rownames(assoc_1se_1M)[idx][which.min(p.adjust(assoc_1se_1M$EA_p[idx]))]
qqplot(assoc_1se_1M$BEA_p[idx])
min(p.adjust(assoc_1se_1M$BEA_p[idx]),na.rm=T)
rownames(assoc_1se_1M)[idx][which.min(p.adjust(assoc_1se_1M$BEA_p[idx]))]




idx1=which(res_1se_500k$glmflag==1)
idx2=which(res_1se_1M$glmflag==1)
length(intersect(idx1,idx2))
idx=intersect(idx1,idx2)
sum(res_1se_1M$r2[idx]>res_1se_500k$r2[idx])
sum(res_1se_1M$selectedsnps[idx]==res_1se_500k$selectedsnps[idx])

load(paste0(outfolder,"/bca_extractgenotype.RData"))
pvalue_assoc_gene=function(gene="DDX49",idx1,idx2)
{
  bcagenotype1=bcagenotype
  # idx=grepl("\\w*_\\w*_\\w*",colnames(bcagenotype1)) #remove those NA samples with more than 1 _
  # bcagenotype1=bcagenotype1[,!idx]
  # tmp=unlist(strsplit(colnames(bcagenotype1),"_"))
  tmp=rep(NA,ncol(bcagenotype1))
  for (i in 1:length(tmp))
  {
    tmp1=unlist(strsplit(colnames(bcagenotype1)[i],"_"))
    tmp[i]=paste0(tmp1[2:length(tmp1)],collapse = "_")
  }
  colnames(bcagenotype1)=tmp
  idx=match(colnames(predict_bcageneexp),colnames(bcagenotype1))
  bcagenotype1=bcagenotype1[,idx]
  tmp=compute_cor_arow(i=which(rownames(res_1se_500k)==gene),opt="1se",ncv=10,distcutoff = 5e5)
  selectedsnps=unlist(strsplit(tmp$selectedsnps,"|",fixed=T))
  pvalues=rep(NA,length(selectedsnps))
  names(pvalues)=selectedsnps
  y=c(rep(1,length(idx1)),rep(0,length(idx2)))
  for (i in 1:length(selectedsnps))
  {
    x=as.numeric(bcagenotype1[which(rownames(bcagenotype1)==selectedsnps[i]),c(idx1,idx2)])
    fm=glm(y~x+age+sex+pc1+pc2+pc3,data=covariates,family=binomial)
    #fm=glm(y~x+age+sex+site1+pc1+pc2+pc3,data=covariates,family=binomial) #including site
    tmp=summary(fm)$coefficients
    if ("x" %in% rownames(tmp))
    {
      pvalues[i]=summary(fm)$coefficients[2,4]
    }
  }
  return(pvalues)
}

computeeQTL=function(selectedsnps,gene="DDX49")
{
  res=rep(NA,length(selectedsnps))
  j=which(rownames(phenotypepos)==gene)
  Y=unlist(phenotype[j,])
  for (i in 1:length(selectedsnps))
  {
    X=t(snp[which(rownames(snp)==selectedsnps[i]),])
    mutationV=unlist(mutation[j,])
    if (all(mutationV==0)) #no mutation
    {
      Xall=data.matrix(cbind(X,cn=unlist(copynumber[j,]),covariate))
      covariateall=data.matrix(cbind(cn=unlist(copynumber[j,]),covariate))
    }else
    {
      Xall=data.matrix(cbind(X,cn=unlist(copynumber[j,]),mutation=mutationV,covariate))
      covariateall=data.matrix(cbind(cn=unlist(copynumber[j,]),mutation=mutationV,covariate))
    }
    fit1=lm(Y~Xall)
    res[i]=summary(fit1)$coefficients[2,4]
  }
}
idx=res_1se_500k$glmflag==1
quantile(res_1se_500k$r2[idx])
quantile(res_1se_500k$numselectedsnp[idx])
idx=res_1se_1M$glmflag==1
quantile(res_1se_1M$r2[idx])
quantile(res_1se_1M$numselectedsnp[idx])

load("../result/Dong23SNPs.RData")
checksnps=c("rs199620551","rs10419226","rs10423674")
knowndat=allgenotypedat[5:nrow(allgenotypedat),which(colnames(allgenotypedat) %in% checksnps)]
# tmp=rep(NA,nrow(knowndat))
# for (i in 1:length(tmp))
# {
#   tmp1=unlist(strsplit(rownames(knowndat)[i],"_"))
#   tmp[i]=paste0(tmp1[2:length(tmp1)],collapse = "_")
# }
# rownames(knowndat)=tmp
seldat=bcagenotype[which(rownames(bcagenotype) %in% selectedsnps),]
seldat=as.data.frame(t(seldat))
idx=match(rownames(seldat),rownames(knowndat))
knowndat=knowndat[idx,]
test=cor(data.matrix(cbind(seldat,knowndat)))
knowndat1=read.table("/fh/fast/dai_j/BEACON/BEACON_GRANT/data/imputation/bca_1000g/chr19.3knownsnps.raw",header=T,stringsAsFactors = F)
rownames(knowndat1)=knowndat1$IID
knowndat1=knowndat1[,7:9]
idx=match(rownames(seldat),rownames(knowndat1))
test1=cor(data.matrix(cbind(seldat,knowndat1)))
write.csv(test1,file="../result/Cormatrix_4selectedSNPS_3knownSNPS_DDX49.csv")
alldat=data.matrix(cbind(seldat,knowndat1))
for (i in 1:ncol(alldat))
{
  print((sum(alldat[,i]==1)+sum(alldat[,i]==2)*2)/2/nrow(alldat))
}
knowndat2=knowndat1
tmp=rep(NA,nrow(knowndat2))
for (i in 1:length(tmp))
{
  tmp1=unlist(strsplit(rownames(knowndat2)[i],"_"))
  tmp[i]=paste0(tmp1[2:length(tmp1)],collapse = "_")
}
rownames(knowndat2)=tmp
idx=match(geneexpsamplenames,rownames(knowndat2))
knowndat2=knowndat2[idx,]
y=c(rep(1,length(idx1)),rep(0,length(idx2)))
x=as.numeric(predict_bcageneexp[idxgene,c(idx1,idx2)])
colnames(knowndat2)=c("snp1","snp2","snp3")
fm=glm(y~x+age+sex+pc1+pc2+pc3+snp1+snp2+snp3,data=cbind.data.frame(covariates,knowndat2[c(idx1,idx2),]),family=binomial)
fm=glm(y~x+age+sex+pc1+pc2+pc3,data=cbind.data.frame(covariates,knowndat2[c(idx1,idx2),]),family=binomial)
fm=glm(y~snp1+age+sex+pc1+pc2+pc3,data=cbind.data.frame(covariates,knowndat2[c(idx1,idx2),]),family=binomial)
fm=glm(y~snp2+age+sex+pc1+pc2+pc3,data=cbind.data.frame(covariates,knowndat2[c(idx1,idx2),]),family=binomial)
fm=glm(y~snp3+age+sex+pc1+pc2+pc3,data=cbind.data.frame(covariates,knowndat2[c(idx1,idx2),]),family=binomial)

gene="CRTC1"
knownsnps=colnames(knowndat1)
knownsnps=gsub("X","",knownsnps)
knownsnps=gsub(".",":",knownsnps,fixed = T)

TCGAknownsnps=read.table("/fh/fast/dai_j/CancerGenomics/EAprogression/result/chr19_3knownsnps.txt",stringsAsFactors = F)
TCGAknownsnps1=data.frame(matrix(NA,nrow=185,ncol=3))
colnames(TCGAknownsnps1)=c("rs199620551","rs10419226","rs10423674")
tmp=read.table("/fh/fast/dai_j/CancerGenomics/EAprogression/result/imputationJune2014/plink/TCGAtumors_chr19_flip.fam",stringsAsFactors = F)
rownames(TCGAknownsnps1)=tmp$V1
for (i in 1:185)
{
  tmp=TCGAknownsnps[,5+(i-1)*3+2]+2*TCGAknownsnps[,5+(i-1)*3+3]
  TCGAknownsnps1[i,]=tmp
}
idx=match(colnames(phenotype),rownames(TCGAknownsnps1))
TCGAknownsnps1=TCGAknownsnps1[idx,]
Y=unlist(phenotype[which(rownames(phenotype)==gene),]) 
mutationV=unlist(mutation[which(rownames(phenotype)==gene),])
covariateall=data.matrix(cbind(cn=unlist(copynumber[which(rownames(phenotype)==gene),]),mutation=mutationV,covariate))
for (i in 1:3)
{
  X1=cbind.data.frame(x=TCGAknownsnps1[,i],covariateall)
  fm=glm(as.formula("Y~."),data=X1)
  print(summary(fm)$coefficients[2,4])
}
which(rownames(snp) %in% c("19:18803172_T","19:18804294_TG","19:18817903_A"))
rownames(snp)[which(rownames(snp) %in% c("19:18803172_T","19:18804294_TG","19:18817903_A"))] #"19:18803172_T"
X1=cbind.data.frame(x=unlist(snp[6035350,]),covariateall)
which(rownames(res_1se_500k)==gene)
res_1se_500k[4266,]
res_1se_1M[4266,]
