#To compute PRS on BCA samples using Validation_EA_genotyped.weight generated by PRS_LDpred_genotype_cambridge.sh

library(data.table)

EAmeta=as.data.frame(fread("../result/Discovery_Bonn_METAANALYSIS_EA_genotyped_comsnp1.tbl"))
colnames(EAmeta)[which(colnames(EAmeta)=="P-value")]="P"
EABeacon=as.data.frame(fread("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/EA_Discovery_genotyped_BD_autosomes_comsnp_N.txt"))
idx=match(EAmeta$MarkerName,EABeacon$SNP)
EAmeta$chr=EABeacon$CHR[idx]
EAmeta$pos=EABeacon$position[idx]

generate_plink=function(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/bca_filtered_30Nov2018_flip_noambiguous",
                        opt="EA",prefix1="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/BCA_EA_genotyped",metadat=EAmeta)
{
  #
  bim <- read.table(paste0(prefix,".bim"),stringsAsFactors = F)
  colnames(bim) <- c("CHR", "SNP", "CM", "BP", "B.A1", "B.A2")
  metadat$Allele1=toupper(metadat$Allele1) #Allele1 is non-effect
  metadat$Allele2=toupper(metadat$Allele2)
  metadat$A2=metadat$Allele1
  metadat$A1=metadat$Allele2
  colnames(metadat)[which(colnames(metadat)=="MarkerName")]="SNP"
  colnames(metadat)[which(colnames(metadat)=="chr")]="CHR"
  colnames(metadat)[which(colnames(metadat)=="pos")]="BP"
  #update bim SNP ID based on metadat
  bim_posname1=paste0(bim$CHR,":",bim$BP,"_",bim$B.A1,"_",bim$B.A2)
  bim_posname2=paste0(bim$CHR,":",bim$BP,"_",bim$B.A2,"_",bim$B.A1)
  meta_posname=paste0(metadat$CHR,":",metadat$BP,"_",metadat$Allele1,"_",metadat$Allele2)
  idx1=which(bim_posname1 %in% meta_posname)
  idx2=which(bim_posname2 %in% meta_posname)
  tmp=bim_posname1[idx1]
  idx3=match(tmp,bim_posname1)
  idx4=match(tmp,meta_posname)
  bim$SNP[idx3]=metadat$SNP[idx4]
  tmp=bim_posname2[idx2]
  idx3=match(tmp,bim_posname2)
  idx4=match(tmp,meta_posname)
  bim$SNP[idx3]=metadat$SNP[idx4]
  print(sum(bim$SNP %in% metadat$SNP))#number of overlapped SNPs
  write.table(bim,file=paste0(prefix1,".bim"),col.names = F,row.names = F,sep=" ",quote=F)
  
  
  for (ext in c(".bed",".fam"))
  {
    #file.copy(paste0(prefix,ext),paste0(prefix1,ext))
    cmd=paste0("cp ",paste0(prefix,ext)," ",paste0(prefix1,ext))
    system(cmd,wait = T)
  }
  fam=read.table(paste0(prefix1,".fam"))
  idx=match(fam$V2,sampletable$localid)
  if (opt=="BE")
  {
    fam$V6=sampletable$phenoBE_bca[idx]
  }
  if (opt=="EA")
  {
    fam$V6=sampletable$phenoEA_bca[idx]
  }
  if (opt=="BEEA")
  {
    fam$V6=sampletable$phenoEABE_bca[idx]
  }
  write.table(fam,file=paste0(prefix1,".fam"),row.names = F,col.names = F,sep=" ",quote=F)
  #generate phenotype file
  tmp=fam[,c(1,2,6)]
  write.table(tmp,file=paste0(prefix1,".pheno"),row.names = F,col.names = F,sep=" ",quote=F)
}

generate_plink()

generate_covariate=function(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/BCA_EA_genotyped")
{
  fam=read.table(paste0(prefix,".fam"),stringsAsFactors = F)
  idx=match(fam$V2,sampletable$localid)
  Covariate=sampletable[idx,colnames(sampletable) %in% c("sex","age","ev1_bca","ev2_bca","ev3_bca","ev4_bca")]
  colnames(Covariate)[which(colnames(Covariate)=="ev1_bca")]="pc1"
  colnames(Covariate)[which(colnames(Covariate)=="ev2_bca")]="pc2"
  colnames(Covariate)[which(colnames(Covariate)=="ev3_bca")]="pc3"
  colnames(Covariate)[which(colnames(Covariate)=="ev4_bca")]="pc4"
  rownames(Covariate)=sampletable$localid[idx]
  Covariate$age[is.na(Covariate$age)]=mean(Covariate$age,na.rm = T)
  Covariate=data.frame(IID=rownames(Covariate),Covariate)
  write.table(Covariate,file=paste0(prefix,".covariate"),col.names = T,row.names = F,sep=" ",quote=F)
}
generate_covariate()

#check results------
library(pROC)
sampletable=readxl::read_excel("/fh/fast/dai_j/BEACON/BEACON_GRANT/data/PLINKinputCombo_bca_07Feb2018.xls",1)
sampletable <- data.frame(sampletable)
plot_ROC=function(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/Beacon_BE_genotyped",opt=1)
{
  phenotype <- read.table(paste0(prefix,".pheno"), header=F)
  colnames(phenotype)=c("FID","IID","case")
  if (opt==1)
  {
    for(i in p.threshold)
    {
      # Go through each p-value threshold .score_LDpred_p1.0000e-01.txt
      tmp=formatC(i, format = "e", digits =4)
      LDpredfile=paste0(prefix,".score_LDpred_p",tmp,".txt")
      if (file.exists(LDpredfile))
      {
        prs <- read.table(LDpredfile, header=T,sep=",",stringsAsFactors = F)
        pheno.prs <- merge(phenotype, prs, by=c("IID"))
        pheno.prs$true_phens[pheno.prs$true_phens==-9]=NA
        pheno.prs$case[pheno.prs$case==-9]=NA
        idx=match(pheno.prs$IID,sampletable$localid)
        sampletable1=sampletable[idx,]
        sampletable1$prs=pheno.prs$PRS
        sampletable1$case=pheno.prs$case
        fit1 <- glm(I(case==2)~prs,family=binomial,data=sampletable1[sampletable1$site<=30,],y=T)
        #testroc<- roc(pheno.prs$true_phens, pheno.prs$PRS)
        
        fit3 <- glm(I(case==2)~age+sex+recurrent_HB_RF+bmi_recent_healthy+cig_smk_ever+nsaid_ever+prs,family=binomial,data=sampletable1[sampletable1$site<=30,],y=T)
        roc1=roc(fit1$y,fit1$fitted.values)
        roc3=roc(fit3$y,fit3$fitted.values)
        print(paste0("p=",i,":auc_prs=",round(roc1$auc,3)," auc_env=",round(roc3$auc,3)))
        #plot(testroc, print.auc=TRUE,main=tmp)
      }
    }
  }
  
  if (opt==2) #LDpred-inf
  {
    LDpredfile=paste0(prefix,".score_LDpred-inf.txt")
    prs <- read.table(LDpredfile, header=T,sep=",",stringsAsFactors = F)
    pheno.prs <- merge(phenotype, prs, by=c("IID"))
    testroc<- roc(pheno.prs$true_phens, pheno.prs$PRS)
    plot(testroc, print.auc=TRUE,main="inf")
  }
  
}

plot_ROC(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/BCA_EA")
# [1] "p=1:auc_prs=0.896 auc_env=0.99"
# [1] "p=0.3:auc_prs=0.896 auc_env=0.99"
# [1] "p=0.1:auc_prs=0.896 auc_env=0.99"
# [1] "p=0.03:auc_prs=0.896 auc_env=0.99"
# [1] "p=0.01:auc_prs=0.895 auc_env=0.99"
# [1] "p=0.003:auc_prs=0.881 auc_env=0.984"
# [1] "p=0.001:auc_prs=0.836 auc_env=0.964"
plot_ROC(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/BCA_EA_genotyped")
# [1] "p=0.3:auc_prs=0.902 auc_env=0.993"
# [1] "p=0.1:auc_prs=0.902 auc_env=0.993"
# [1] "p=0.03:auc_prs=0.902 auc_env=0.993"
# [1] "p=0.01:auc_prs=0.896 auc_env=0.991"
# [1] "p=0.003:auc_prs=0.867 auc_env=0.98"
# [1] "p=0.001:auc_prs=0.811 auc_env=0.954"

#calculate ordds ratios from PRS
require(MASS)
#PRS high, risk high, for genotypted
PRS_odds=function(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/Beacon_BE",p=0.1)
{
  phenotype <- read.table(paste0(prefix,".pheno"), header=F)
  colnames(phenotype)=c("FID","IID","case")
  
  tmp=formatC(p, format = "e", digits =4)
  LDpredfile=paste0(prefix,".score_LDpred_p",tmp,".txt.adj")
  if (file.exists(LDpredfile))
  {
    prs <- read.table(LDpredfile, header=T,sep=",")
    pheno.prs <- merge(phenotype, prs, by=c("IID"))
    #quantiles:
    qt=quantile(pheno.prs$PRS)
    qtidx=cut(pheno.prs$PRS,qt)
    pheno.prs$qtidx=qtidx
    pheno.prs$group=as.numeric(qtidx)
    pheno.prs$true_phens[pheno.prs$true_phens==-9]=NA
    pheno.prs$case[pheno.prs$case==-9]=NA
    if (sum(pheno.prs$case==2,na.rm = T)>0)
    {
      pheno.prs$case[which(pheno.prs$case==1)]=0
      pheno.prs$case[which(pheno.prs$case==2)]=1
    }
    tmp=pheno.prs[,c("IID","PRS")]
    write.table(tmp,file=paste0(prefix,".PRS.txt"),sep="\t",row.names = F,quote=F)
    for (i in 2:4)
    {
      print(paste0(i,"th quantile:"))
      dat=pheno.prs[pheno.prs$group %in% c(1,i),]
      dat$group=factor(dat$group,levels = c(1,i))
      fit=glm(I(dat$case==1)~factor(dat$group),family = binomial)
      print(exp(cbind(coef(fit), confint(fit)))[2,])
    }
    #return(tmp)
  }
}

PRS_odds(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/BCA_EA_genotyped",p=0.1)
PRS_odds(prefix="/fh/fast/dai_j/BEACON/BEACON_GRANT/result/BCA_EA",p=0.01)

#check two sets of PRS
readPRS=function(prefix="Validation_EA")
{
  EAfile=paste0("/fh/fast/dai_j/BEACON/BEACON_GRANT/result/",prefix,".PRS.txt")
  EAprs=read.table(EAfile,header=T,stringsAsFactors = F)
  names(EAprs) <- c("localid","EA.prs")
  return(EAprs)
}
#used in validation
CambridgePRS=readPRS()
BCAPRS=readPRS(prefix="BCA_EA")
tmp=merge(CambridgePRS,BCAPRS,by="localid")
cor(tmp$EA.prs.x,tmp$EA.prs.y) #1
plot(tmp$EA.prs.x,tmp$EA.prs.y,xlab="oldPRS",ylab="newPRS",cex.lab=1.3,cex.axis=1.3)
abline(0,1)
